// vscode/vscode/src/vs/workbench/browser/window.js
import { __decorate, __param } from '../../../../../external/tslib/tslib.es6.js';
import { setFullscreen, isSafari } from '../../base/browser/browser.js';
import { getWindowsCount, getWindows, getActiveWindow, getWindow, getWindowById, addDisposableListener, EventType, EventHelper, windowOpenWithSuccess, windowOpenPopup, windowOpenNoOpener } from '../../base/browser/dom.js';
import { DomEmitter } from '../../base/browser/event.js';
import { requestUsbDevice, requestSerialPort, requestHidDevice } from '../../base/browser/deviceAccess.js';
import { timeout } from '../../base/common/async.js';
import { Event } from '../../base/common/event.js';
import { Disposable, dispose, toDisposable } from '../../base/common/lifecycle.js';
import { matchesScheme, Schemas } from '../../base/common/network.js';
import { isMacintosh, isIOS } from '../../base/common/platform.js';
import Severity from '../../base/common/severity.js';
import { URI } from '../../base/common/uri.js';
import { localizeWithPath } from '../../nls.js';
import { CommandsRegistry } from '../../platform/commands/common/commands.js';
import '../../../../../override/vs/platform/dialogs/common/dialogs.js';
import { IInstantiationService } from '../../platform/instantiation/common/instantiation.js';
import { ILabelService } from '../../platform/label/common/label.js';
import { IOpenerService } from '../../platform/opener/common/opener.js';
import { IProductService } from '../../platform/product/common/productService.js';
import { IBrowserWorkbenchEnvironmentService } from '../services/environment/browser/environmentService.js';
import { IWorkbenchLayoutService } from '../services/layout/browser/layoutService.js';
import { ILifecycleService } from '../services/lifecycle/common/lifecycle.js';
import { IHostService } from '../services/host/browser/host.js';
import { registerWindowDriver } from '../services/driver/browser/driver.js';
import { isAuxiliaryWindow, mainWindow } from '../../base/browser/window.js';
import { createSingleCallFunction } from '../../base/common/functional.js';
import { IConfigurationService } from '../../platform/configuration/common/configuration.js';
import { IDialogService } from '../../platform/dialogs/common/dialogs.js';
var BaseWindow_1;
let BaseWindow = class BaseWindow extends Disposable {
    static { BaseWindow_1 = this; }
    static { this.TIMEOUT_HANDLES = Number.MIN_SAFE_INTEGER; }
    static { this.TIMEOUT_DISPOSABLES = ( new Map()); }
    constructor(targetWindow, dom = { getWindowsCount, getWindows }, hostService) {
        super();
        this.hostService = hostService;
        this.enableWindowFocusOnElementFocus(targetWindow);
        this.enableMultiWindowAwareTimeout(targetWindow, dom);
        this.registerFullScreenListeners(targetWindow.vscodeWindowId);
    }
    enableWindowFocusOnElementFocus(targetWindow) {
        const originalFocus = HTMLElement.prototype.focus;
        targetWindow.HTMLElement.prototype.focus = function (options) {
            const activeWindow = getActiveWindow();
            if (activeWindow.document.hasFocus()) {
                const elementWindow = getWindow(this);
                if (activeWindow !== elementWindow) {
                    elementWindow.focus();
                }
            }
            originalFocus.apply(this, [options]);
        };
    }
    enableMultiWindowAwareTimeout(targetWindow, dom = { getWindowsCount, getWindows }) {
        const originalSetTimeout = targetWindow.setTimeout;
        Object.defineProperty(targetWindow, 'vscodeOriginalSetTimeout', { get: () => originalSetTimeout });
        const originalClearTimeout = targetWindow.clearTimeout;
        Object.defineProperty(targetWindow, 'vscodeOriginalClearTimeout', { get: () => originalClearTimeout });
        targetWindow.setTimeout = function (handler, timeout = 0, ...args) {
            if (dom.getWindowsCount() === 1 || typeof handler === 'string' || timeout === 0 ) {
                return originalSetTimeout.apply(this, [handler, timeout, ...args]);
            }
            const timeoutDisposables = ( new Set());
            const timeoutHandle = BaseWindow_1.TIMEOUT_HANDLES++;
            BaseWindow_1.TIMEOUT_DISPOSABLES.set(timeoutHandle, timeoutDisposables);
            const handlerFn = createSingleCallFunction(handler, () => {
                dispose(timeoutDisposables);
                BaseWindow_1.TIMEOUT_DISPOSABLES.delete(timeoutHandle);
            });
            for (const { window, disposables } of dom.getWindows()) {
                if (isAuxiliaryWindow(window) && window.document.visibilityState === 'hidden') {
                    continue;
                }
                const handle = window.vscodeOriginalSetTimeout.apply(this, [handlerFn, timeout, ...args]);
                const timeoutDisposable = toDisposable(() => {
                    window.vscodeOriginalClearTimeout(handle);
                    timeoutDisposables.delete(timeoutDisposable);
                });
                disposables.add(timeoutDisposable);
                timeoutDisposables.add(timeoutDisposable);
            }
            return timeoutHandle;
        };
        targetWindow.clearTimeout = function (timeoutHandle) {
            const timeoutDisposables = typeof timeoutHandle === 'number' ? BaseWindow_1.TIMEOUT_DISPOSABLES.get(timeoutHandle) : undefined;
            if (timeoutDisposables) {
                dispose(timeoutDisposables);
                BaseWindow_1.TIMEOUT_DISPOSABLES.delete(timeoutHandle);
            }
            else {
                originalClearTimeout.apply(this, [timeoutHandle]);
            }
        };
    }
    registerFullScreenListeners(targetWindowId) {
        this._register(this.hostService.onDidChangeFullScreen(({ windowId, fullscreen }) => {
            if (windowId === targetWindowId) {
                const targetWindow = getWindowById(targetWindowId);
                if (targetWindow) {
                    setFullscreen(fullscreen, targetWindow.window);
                }
            }
        }));
    }
    static async confirmOnShutdown(accessor, reason) {
        const dialogService = accessor.get(IDialogService);
        const configurationService = accessor.get(IConfigurationService);
        const message = reason === 2  ?
            (isMacintosh ? ( localizeWithPath(
            'vs/workbench/browser/window',
            'quitMessageMac',
            "Are you sure you want to quit?"
        )) : ( localizeWithPath(
            'vs/workbench/browser/window',
            'quitMessage',
            "Are you sure you want to exit?"
        ))) :
            ( localizeWithPath(
                'vs/workbench/browser/window',
                'closeWindowMessage',
                "Are you sure you want to close the window?"
            ));
        const primaryButton = reason === 2  ?
            (isMacintosh ? ( localizeWithPath(
            'vs/workbench/browser/window',
            { key: 'quitButtonLabel', comment: ['&& denotes a mnemonic'] },
            "&&Quit"
        )) : ( localizeWithPath(
            'vs/workbench/browser/window',
            { key: 'exitButtonLabel', comment: ['&& denotes a mnemonic'] },
            "&&Exit"
        ))) :
            ( localizeWithPath(
                'vs/workbench/browser/window',
                { key: 'closeWindowButtonLabel', comment: ['&& denotes a mnemonic'] },
                "&&Close Window"
            ));
        const res = await dialogService.confirm({
            message,
            primaryButton,
            checkbox: {
                label: ( localizeWithPath('vs/workbench/browser/window', 'doNotAskAgain', "Do not ask me again"))
            }
        });
        if (res.confirmed && res.checkboxChecked) {
            await configurationService.updateValue('window.confirmBeforeClose', 'never');
        }
        return res.confirmed;
    }
};
BaseWindow = BaseWindow_1 = ( __decorate([
    ( __param(2, IHostService))
], BaseWindow));
let BrowserWindow = class BrowserWindow extends BaseWindow {
    constructor(openerService, lifecycleService, dialogService, labelService, productService, environmentService, layoutService, instantiationService, hostService) {
        super(mainWindow, undefined, hostService);
        this.openerService = openerService;
        this.lifecycleService = lifecycleService;
        this.dialogService = dialogService;
        this.labelService = labelService;
        this.productService = productService;
        this.environmentService = environmentService;
        this.layoutService = layoutService;
        this.instantiationService = instantiationService;
        this.registerListeners();
        this.create();
    }
    registerListeners() {
        this._register(this.lifecycleService.onWillShutdown(() => this.onWillShutdown()));
        const viewport = isIOS && mainWindow.visualViewport ? mainWindow.visualViewport  : mainWindow ;
        this._register(addDisposableListener(viewport, EventType.RESIZE, () => {
            this.layoutService.layout();
            if (isIOS) {
                mainWindow.scrollTo(0, 0);
            }
        }));
        this._register(addDisposableListener(this.layoutService.mainContainer, EventType.WHEEL, e => e.preventDefault(), { passive: false }));
        this._register(addDisposableListener(this.layoutService.mainContainer, EventType.CONTEXT_MENU, e => EventHelper.stop(e, true)));
        this._register(addDisposableListener(this.layoutService.mainContainer, EventType.DROP, e => EventHelper.stop(e, true)));
    }
    onWillShutdown() {
        Event.toPromise(Event.any(Event.once(( new DomEmitter(mainWindow.document.body, EventType.KEY_DOWN, true)).event), Event.once(( new DomEmitter(mainWindow.document.body, EventType.MOUSE_DOWN, true)).event))).then(async () => {
            await timeout(3000);
            await this.dialogService.prompt({
                type: Severity.Error,
                message: ( localizeWithPath(
                    'vs/workbench/browser/window',
                    'shutdownError',
                    "An unexpected error occurred that requires a reload of this page."
                )),
                detail: ( localizeWithPath(
                    'vs/workbench/browser/window',
                    'shutdownErrorDetail',
                    "The workbench was unexpectedly disposed while running."
                )),
                buttons: [
                    {
                        label: ( localizeWithPath(
                            'vs/workbench/browser/window',
                            { key: 'reload', comment: ['&& denotes a mnemonic'] },
                            "&&Reload"
                        )),
                        run: () => mainWindow.location.reload()
                    }
                ]
            });
        });
    }
    create() {
        this.setupOpenHandlers();
        this.registerLabelFormatters();
        this.registerCommands();
        this.setupDriver();
    }
    setupDriver() {
        if (this.environmentService.enableSmokeTestDriver) {
            registerWindowDriver(this.instantiationService);
        }
    }
    setupOpenHandlers() {
        this.openerService.setDefaultExternalOpener({
            openExternal: async (href) => {
                let isAllowedOpener = false;
                if (this.environmentService.options?.openerAllowedExternalUrlPrefixes) {
                    for (const trustedPopupPrefix of this.environmentService.options.openerAllowedExternalUrlPrefixes) {
                        if (href.startsWith(trustedPopupPrefix)) {
                            isAllowedOpener = true;
                            break;
                        }
                    }
                }
                if (matchesScheme(href, Schemas.http) || matchesScheme(href, Schemas.https)) {
                    if (isSafari) {
                        const opened = windowOpenWithSuccess(href, !isAllowedOpener);
                        if (!opened) {
                            await this.dialogService.prompt({
                                type: Severity.Warning,
                                message: ( localizeWithPath(
                                    'vs/workbench/browser/window',
                                    'unableToOpenExternal',
                                    "The browser interrupted the opening of a new tab or window. Press 'Open' to open it anyway."
                                )),
                                detail: href,
                                buttons: [
                                    {
                                        label: ( localizeWithPath(
                                            'vs/workbench/browser/window',
                                            { key: 'open', comment: ['&& denotes a mnemonic'] },
                                            "&&Open"
                                        )),
                                        run: () => isAllowedOpener ? windowOpenPopup(href) : windowOpenNoOpener(href)
                                    },
                                    {
                                        label: ( localizeWithPath(
                                            'vs/workbench/browser/window',
                                            { key: 'learnMore', comment: ['&& denotes a mnemonic'] },
                                            "&&Learn More"
                                        )),
                                        run: () => this.openerService.open(( URI.parse('https://aka.ms/allow-vscode-popup')))
                                    }
                                ],
                                cancelButton: true
                            });
                        }
                    }
                    else {
                        isAllowedOpener
                            ? windowOpenPopup(href)
                            : windowOpenNoOpener(href);
                    }
                }
                else {
                    const invokeProtocolHandler = () => {
                        this.lifecycleService.withExpectedShutdown({ disableShutdownHandling: true }, () => mainWindow.location.href = href);
                    };
                    invokeProtocolHandler();
                    const showProtocolUrlOpenedDialog = async () => {
                        const { downloadUrl } = this.productService;
                        let detail;
                        const buttons = [
                            {
                                label: ( localizeWithPath(
                                    'vs/workbench/browser/window',
                                    { key: 'openExternalDialogButtonRetry.v2', comment: ['&& denotes a mnemonic'] },
                                    "&&Try Again"
                                )),
                                run: () => invokeProtocolHandler()
                            }
                        ];
                        if (downloadUrl !== undefined) {
                            detail = ( localizeWithPath(
                                'vs/workbench/browser/window',
                                'openExternalDialogDetail.v2',
                                "We launched {0} on your computer.\n\nIf {1} did not launch, try again or install it below.",
                                this.productService.nameLong,
                                this.productService.nameLong
                            ));
                            buttons.push({
                                label: ( localizeWithPath(
                                    'vs/workbench/browser/window',
                                    { key: 'openExternalDialogButtonInstall.v3', comment: ['&& denotes a mnemonic'] },
                                    "&&Install"
                                )),
                                run: async () => {
                                    await this.openerService.open(( URI.parse(downloadUrl)));
                                    showProtocolUrlOpenedDialog();
                                }
                            });
                        }
                        else {
                            detail = ( localizeWithPath(
                                'vs/workbench/browser/window',
                                'openExternalDialogDetailNoInstall',
                                "We launched {0} on your computer.\n\nIf {1} did not launch, try again below.",
                                this.productService.nameLong,
                                this.productService.nameLong
                            ));
                        }
                        await this.hostService.withExpectedShutdown(() => this.dialogService.prompt({
                            type: Severity.Info,
                            message: ( localizeWithPath(
                                'vs/workbench/browser/window',
                                'openExternalDialogTitle',
                                "All done. You can close this tab now."
                            )),
                            detail,
                            buttons,
                            cancelButton: true
                        }));
                    };
                    if (matchesScheme(href, this.productService.urlProtocol)) {
                        await showProtocolUrlOpenedDialog();
                    }
                }
                return true;
            }
        });
    }
    registerLabelFormatters() {
        this._register(this.labelService.registerFormatter({
            scheme: Schemas.vscodeUserData,
            priority: true,
            formatting: {
                label: '(Settings) ${path}',
                separator: '/',
            }
        }));
    }
    registerCommands() {
        CommandsRegistry.registerCommand('workbench.experimental.requestUsbDevice', async (_accessor, options) => {
            return requestUsbDevice(options);
        });
        CommandsRegistry.registerCommand('workbench.experimental.requestSerialPort', async (_accessor, options) => {
            return requestSerialPort(options);
        });
        CommandsRegistry.registerCommand('workbench.experimental.requestHidDevice', async (_accessor, options) => {
            return requestHidDevice(options);
        });
    }
};
BrowserWindow = ( __decorate([
    ( __param(0, IOpenerService)),
    ( __param(1, ILifecycleService)),
    ( __param(2, IDialogService)),
    ( __param(3, ILabelService)),
    ( __param(4, IProductService)),
    ( __param(5, IBrowserWorkbenchEnvironmentService)),
    ( __param(6, IWorkbenchLayoutService)),
    ( __param(7, IInstantiationService)),
    ( __param(8, IHostService))
], BrowserWindow));
export { BaseWindow, BrowserWindow };
