// vscode/vscode/src/vs/base/browser/ui/findinput/findInputToggles.js
import { Toggle } from '../toggle/toggle.js';
import { Codicon } from '../../../common/codicons.js';
import { localizeWithPath } from '../../../../nls.js';
const NLS_CASE_SENSITIVE_TOGGLE_LABEL = ( localizeWithPath(
    'vs/base/browser/ui/findinput/findInputToggles',
    'caseDescription',
    "Match Case"
));
const NLS_WHOLE_WORD_TOGGLE_LABEL = ( localizeWithPath(
    'vs/base/browser/ui/findinput/findInputToggles',
    'wordsDescription',
    "Match Whole Word"
));
const NLS_REGEX_TOGGLE_LABEL = ( localizeWithPath(
    'vs/base/browser/ui/findinput/findInputToggles',
    'regexDescription',
    "Use Regular Expression"
));
class CaseSensitiveToggle extends Toggle {
    constructor(opts) {
        super({
            icon: Codicon.caseSensitive,
            title: NLS_CASE_SENSITIVE_TOGGLE_LABEL + opts.appendTitle,
            isChecked: opts.isChecked,
            inputActiveOptionBorder: opts.inputActiveOptionBorder,
            inputActiveOptionForeground: opts.inputActiveOptionForeground,
            inputActiveOptionBackground: opts.inputActiveOptionBackground
        });
    }
}
class WholeWordsToggle extends Toggle {
    constructor(opts) {
        super({
            icon: Codicon.wholeWord,
            title: NLS_WHOLE_WORD_TOGGLE_LABEL + opts.appendTitle,
            isChecked: opts.isChecked,
            inputActiveOptionBorder: opts.inputActiveOptionBorder,
            inputActiveOptionForeground: opts.inputActiveOptionForeground,
            inputActiveOptionBackground: opts.inputActiveOptionBackground
        });
    }
}
class RegexToggle extends Toggle {
    constructor(opts) {
        super({
            icon: Codicon.regex,
            title: NLS_REGEX_TOGGLE_LABEL + opts.appendTitle,
            isChecked: opts.isChecked,
            inputActiveOptionBorder: opts.inputActiveOptionBorder,
            inputActiveOptionForeground: opts.inputActiveOptionForeground,
            inputActiveOptionBackground: opts.inputActiveOptionBackground
        });
    }
}
export { CaseSensitiveToggle, RegexToggle, WholeWordsToggle };
