// monaco-editor-wrapper/bundle/index-254ppoW-.js
import { a9 as tt, aa as ot, ab as bo, ac as So, ad as Co, ae as Io, af as yo, ag as wo, ah as ko, ai as vo, aj as Eo, ak as xo, al as Ro, am as _o, an as Do, P as V, ao as J, ap as Wt, aq as Z, ar as Ae, as as kt, Q as Me, at as Po, au as Mo, av as xe, aw as Oo, ax as vt, ay as j, az as w, aA as Ue, aB as Be, aC as ze, aD as Lo, aE as at, aF as Ho, aG as ht, B as r, aH as lt, aI as mt, J as X, aJ as Fe, aK as M, aL as Re, aM as Ao, aN as Fo, aO as dt, aP as _e, aQ as qe, aR as ut, aS as Jt, aT as No, a4 as Yt, z as ft, aU as Ne, D as $o, aV as jo, E as Ce, V as Et, aW as Go, j as gt, aX as v, aY as S, aZ as xt, a_ as nt, a$ as Uo, b0 as Bo, b1 as Je, b2 as zo, b3 as qo, K as $e, G as Qt, b4 as Ko, b5 as Vo, b6 as Wo, b7 as Jo, b8 as Rt, b9 as Yo, ba as Zt, q as Xt, w as eo, bb as Qo, bc as Zo, I as pt, bd as Xo, Y as to, be as en, bf as tn, s as on, bg as nn, bh as _t, bi as sn, bj as rn, bk as he, bl as le, bm as me, bn as de, bo as cn, X as ee, $ as ke, bp as an, bq as hn, br as Dt, bs as ln, bt as Tt, bu as bt, bv as ve, bw as Pt, bx as mn, by as dn, bz as un, bA as fn, bB as gn, bC as pn, bD as oo, Z as Tn, bE as no, bF as bn, v as Sn, bG as Cn, bH as In, a2 as yn, bI as wn } from "./index-8r3crY85.js";
function Ke(s, e, t, o) {
  var n = arguments.length, i = n < 3 ? e : o === null ? o = Object.getOwnPropertyDescriptor(e, t) : o, c;
  if (typeof Reflect == "object" && typeof Reflect.decorate == "function")
    i = Reflect.decorate(s, e, t, o);
  else
    for (var a = s.length - 1; a >= 0; a--)
      (c = s[a]) && (i = (n < 3 ? c(i) : n > 3 ? c(e, t, i) : c(e, t)) || i);
  return n > 3 && i && Object.defineProperty(e, t, i), i;
}
function k(s, e) {
  return function(t, o) {
    e(t, o, s);
  };
}
const it = {};
function D(s, e) {
  let t = it[s];
  t || (it[s] = t = []), t.push(e);
}
function io(s, e) {
  for (const t of s)
    if (e.textMateRules.push(t), !t.scope) {
      const o = t.settings;
      if (!o)
        t.settings = {};
      else
        for (const n in o) {
          const i = n, c = it[i];
          if (c) {
            const a = o[i];
            if (typeof a == "string") {
              const l = V.fromHex(a);
              for (const d of c)
                e.colors[d] = l;
            }
          }
          i !== "foreground" && i !== "background" && i !== "fontStyle" && delete o[i];
        }
    }
}
D("background", tt);
D("foreground", ot);
D("selection", bo);
D("inactiveSelection", So);
D("selectionHighlightColor", Co);
D("findMatchHighlight", Io);
D("currentFindMatchHighlight", yo);
D("hoverHighlight", wo);
D("wordHighlight", "editor.wordHighlightBackground");
D("wordHighlightStrong", "editor.wordHighlightStrongBackground");
D("findRangeHighlight", ko);
D("findMatchHighlight", "peekViewResult.matchHighlightBackground");
D("referenceHighlight", "peekViewEditor.matchHighlightBackground");
D("lineHighlight", vo);
D("rangeHighlight", Eo);
D("caret", xo);
D("invisibles", Ro);
D("guide", _o);
D("activeGuide", Do);
const kn = [
  "ansiBlack",
  "ansiRed",
  "ansiGreen",
  "ansiYellow",
  "ansiBlue",
  "ansiMagenta",
  "ansiCyan",
  "ansiWhite",
  "ansiBrightBlack",
  "ansiBrightRed",
  "ansiBrightGreen",
  "ansiBrightYellow",
  "ansiBrightBlue",
  "ansiBrightMagenta",
  "ansiBrightCyan",
  "ansiBrightWhite"
];
for (const s of kn)
  D(s, "terminal." + s);
function vn(s) {
  return En(s, null, null);
}
function En(s, e, t) {
  const o = s.length;
  let n = 0, i = 1, c = 0;
  o > 0 && s.charCodeAt(0) === 65279 && (n = 1);
  function a(g) {
    if (t === null)
      n = n + g;
    else
      for (; g > 0; )
        s.charCodeAt(n) === 10 ? (n++, i++, c = 0) : (n++, c++), g--;
  }
  function l(g) {
    t === null ? n = g : a(g - n);
  }
  function d() {
    for (; n < o; ) {
      const g = s.charCodeAt(n);
      if (g !== 32 && g !== 9 && g !== 13 && g !== 10)
        break;
      a(1);
    }
  }
  function h(g) {
    return s.substr(n, g.length) === g ? (a(g.length), !0) : !1;
  }
  function u(g) {
    const R = s.indexOf(g, n);
    l(R !== -1 ? R + g.length : o);
  }
  function f(g) {
    const R = s.indexOf(g, n);
    if (R !== -1) {
      const _ = s.substring(n, R);
      return l(R + g.length), _;
    } else {
      const _ = s.substr(n);
      return l(o), _;
    }
  }
  let m = 0, p = null;
  const b = [], C = [];
  let T = null;
  function I(g, R) {
    b.push(m), C.push(p), m = g, p = R;
  }
  function P() {
    if (b.length === 0)
      return y("illegal state stack");
    m = b.pop(), p = C.pop();
  }
  function y(g) {
    throw new Error("Near offset " + n + ": " + g + " ~~~" + s.substr(n, 50) + "~~~");
  }
  const te = {
    enterDict: function() {
      if (T === null)
        return y("missing <key>");
      const g = {};
      t !== null && (g[t] = {
        filename: e,
        line: i,
        char: c
      }), p[T] = g, T = null, I(1, g);
    },
    enterArray: function() {
      if (T === null)
        return y("missing <key>");
      const g = [];
      p[T] = g, T = null, I(2, g);
    }
  }, G = {
    enterDict: function() {
      const g = {};
      t !== null && (g[t] = {
        filename: e,
        line: i,
        char: c
      }), p.push(g), I(1, g);
    },
    enterArray: function() {
      const g = [];
      p.push(g), I(2, g);
    }
  };
  function Y() {
    m === 1 ? te.enterDict() : m === 2 ? G.enterDict() : (p = {}, t !== null && (p[t] = {
      filename: e,
      line: i,
      char: c
    }), I(1, p));
  }
  function H() {
    if (m === 1)
      P();
    else
      return y("unexpected </dict>");
  }
  function fe() {
    m === 1 ? te.enterArray() : m === 2 ? G.enterArray() : (p = [], I(2, p));
  }
  function A() {
    if (m === 1)
      return y("unexpected </array>");
    if (m === 2)
      P();
    else
      return y("unexpected </array>");
  }
  function ge(g) {
    if (m === 1) {
      if (T !== null)
        return y("too many <key>");
      T = g;
    } else
      return y("unexpected <key>");
  }
  function pe(g) {
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function x(g) {
    if (isNaN(g))
      return y("cannot parse float");
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function E(g) {
    if (isNaN(g))
      return y("cannot parse integer");
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function U(g) {
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function B(g) {
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function F(g) {
    if (m === 1) {
      if (T === null)
        return y("missing <key>");
      p[T] = g, T = null;
    } else
      m === 2 ? p.push(g) : p = g;
  }
  function po(g) {
    return g.replace(/&#([0-9]+);/g, function(R, _) {
      return String.fromCodePoint(parseInt(_, 10));
    }).replace(/&#x([0-9a-f]+);/g, function(R, _) {
      return String.fromCodePoint(parseInt(_, 16));
    }).replace(/&amp;|&lt;|&gt;|&quot;|&apos;/g, function(R) {
      switch (R) {
        case "&amp;":
          return "&";
        case "&lt;":
          return "<";
        case "&gt;":
          return ">";
        case "&quot;":
          return '"';
        case "&apos;":
          return "'";
      }
      return R;
    });
  }
  function To() {
    let g = f(">"), R = !1;
    return g.charCodeAt(g.length - 1) === 47 && (R = !0, g = g.substring(0, g.length - 1)), {
      name: g.trim(),
      isClosed: R
    };
  }
  function Q(g) {
    if (g.isClosed)
      return "";
    const R = f("</");
    return u(">"), po(R);
  }
  for (; n < o && (d(), !(n >= o)); ) {
    const g = s.charCodeAt(n);
    if (a(1), g !== 60)
      return y("expected <");
    if (n >= o)
      return y("unexpected end of input");
    const R = s.charCodeAt(n);
    if (R === 63) {
      a(1), u("?>");
      continue;
    }
    if (R === 33) {
      if (a(1), h("--")) {
        u("-->");
        continue;
      }
      u(">");
      continue;
    }
    if (R === 47) {
      if (a(1), d(), h("plist")) {
        u(">");
        continue;
      }
      if (h("dict")) {
        u(">"), H();
        continue;
      }
      if (h("array")) {
        u(">"), A();
        continue;
      }
      return y("unexpected closed tag");
    }
    const _ = To();
    switch (_.name) {
      case "dict":
        Y(), _.isClosed && H();
        continue;
      case "array":
        fe(), _.isClosed && A();
        continue;
      case "key":
        ge(Q(_));
        continue;
      case "string":
        pe(Q(_));
        continue;
      case "real":
        x(parseFloat(Q(_)));
        continue;
      case "integer":
        E(parseInt(Q(_), 10));
        continue;
      case "date":
        U(new Date(Q(_)));
        continue;
      case "data":
        B(Q(_));
        continue;
      case "true":
        Q(_), F(!0);
        continue;
      case "false":
        Q(_), F(!1);
        continue;
    }
    if (!/^plist/.test(_.name))
      return y("unexpected opened tag " + _.name);
  }
  return p;
}
function Mt(s, e, t) {
  const o = xn(s);
  let n = o.next();
  for (; n !== null; ) {
    let l = 0;
    if (n.length === 2 && n.charAt(1) === ":") {
      switch (n.charAt(0)) {
        case "R":
          l = 1;
          break;
        case "L":
          l = -1;
          break;
        default:
          console.log(`Unknown priority ${n} in scope selector`);
      }
      n = o.next();
    }
    const d = c();
    if (d && t.push({ matcher: d, priority: l }), n !== ",")
      break;
    n = o.next();
  }
  function i() {
    if (n === "-") {
      n = o.next();
      const l = i();
      return l ? (d) => l(d) < 0 ? 0 : -1 : null;
    }
    if (n === "(") {
      n = o.next();
      const l = a();
      return n === ")" && (n = o.next()), l;
    }
    if (Ot(n)) {
      const l = [];
      do
        l.push(n), n = o.next();
      while (Ot(n));
      return (d) => e(l, d);
    }
    return null;
  }
  function c() {
    let l = i();
    if (!l)
      return null;
    const d = [];
    for (; l; )
      d.push(l), l = i();
    return (h) => {
      let u = d[0](h);
      for (let f = 1; u >= 0 && f < d.length; f++)
        u = Math.min(u, d[f](h));
      return u;
    };
  }
  function a() {
    let l = c();
    if (!l)
      return null;
    const d = [];
    for (; l && (d.push(l), n === "|" || n === ","); ) {
      do
        n = o.next();
      while (n === "|" || n === ",");
      l = c();
    }
    return (h) => {
      let u = d[0](h);
      for (let f = 1; f < d.length; f++)
        u = Math.max(u, d[f](h));
      return u;
    };
  }
}
function Ot(s) {
  return !!s && !!s.match(/[\w\.:]+/);
}
function xn(s) {
  const e = /([LR]:|[\w\.:][\w\.:\-]*|[\,\|\-\(\)])/g;
  let t = e.exec(s);
  return {
    next: () => {
      if (!t)
        return null;
      const o = t[0];
      return t = e.exec(s), o;
    }
  };
}
const Rn = J.as(Wt.ColorContribution), Oe = Ao(), Lt = {
  comments: ["comment", "punctuation.definition.comment"],
  strings: ["string", "meta.embedded.assembly"],
  keywords: ["keyword - keyword.operator", "keyword.control", "storage", "storage.type"],
  numbers: ["constant.numeric"],
  types: ["entity.name.type", "entity.name.class", "support.type", "support.class"],
  functions: ["entity.name.function", "support.function"],
  variables: ["variable", "entity.name.variable"]
}, $ = class $ {
  constructor(e, t, o) {
    this.themeTokenColors = [], this.customTokenColors = [], this.colorMap = {}, this.customColorMap = {}, this.semanticTokenRules = [], this.customSemanticTokenRules = [], this.textMateThemingRules = void 0, this.tokenColorIndex = void 0, this.id = e, this.label = t, this.settingsId = o, this.isLoaded = !1;
  }
  get semanticHighlighting() {
    return this.customSemanticHighlighting !== void 0 ? this.customSemanticHighlighting : this.customSemanticHighlightingDeprecated !== void 0 ? this.customSemanticHighlightingDeprecated : !!this.themeSemanticHighlighting;
  }
  get tokenColors() {
    if (!this.textMateThemingRules) {
      let i = function(c) {
        c.scope && c.settings && (c.scope === "token.info-token" && (n = !0), e.push({ scope: c.scope, settings: { foreground: re(c.settings.foreground), background: re(c.settings.background), fontStyle: c.settings.fontStyle } }));
      };
      const e = [], t = this.getColor(ot) || this.getDefault(ot), o = this.getColor(tt) || this.getDefault(tt);
      e.push({
        settings: {
          foreground: re(t),
          background: re(o)
        }
      });
      let n = !1;
      this.themeTokenColors.forEach(i), this.customTokenColors.forEach(i), n || Dn[this.type].forEach(i), this.textMateThemingRules = e;
    }
    return this.textMateThemingRules;
  }
  getColor(e, t) {
    let o = this.customColorMap[e];
    return o || (o = this.colorMap[e], t !== !1 && Z(o) && (o = this.getDefault(e)), o);
  }
  getTokenStyle(e, t, o, n = !0, i = {}) {
    const c = {
      foreground: void 0,
      bold: void 0,
      underline: void 0,
      strikethrough: void 0,
      italic: void 0
    }, a = {
      foreground: -1,
      bold: -1,
      underline: -1,
      strikethrough: -1,
      italic: -1
    };
    function l(u, f, m) {
      f.foreground && a.foreground <= u && (a.foreground = u, c.foreground = f.foreground, i.foreground = m);
      for (const p of ["bold", "underline", "strikethrough", "italic"]) {
        const b = p, C = f[b];
        C !== void 0 && a[b] <= u && (a[b] = u, c[b] = C, i[b] = m);
      }
    }
    function d(u) {
      const f = u.selector.match(e, t, o);
      f >= 0 && l(f, u.style, u);
    }
    this.semanticTokenRules.forEach(d), this.customSemanticTokenRules.forEach(d);
    let h = !1;
    for (const u in a) {
      const f = u;
      a[f] === -1 ? h = !0 : a[f] = Number.MAX_VALUE;
    }
    if (h)
      for (const u of Oe.getTokenStylingDefaultRules()) {
        const f = u.selector.match(e, t, o);
        if (f >= 0) {
          let m;
          if (u.defaults.scopesToProbe && (m = this.resolveScopes(u.defaults.scopesToProbe), m && l(f, m, u.defaults.scopesToProbe)), !m && n !== !1) {
            const p = u.defaults[this.type];
            m = this.resolveTokenStyleValue(p), m && l(f, m, p);
          }
        }
      }
    return Ae.fromData(c);
  }
  resolveTokenStyleValue(e) {
    if (e !== void 0) {
      if (typeof e == "string") {
        const { type: t, modifiers: o, language: n } = kt(e, "");
        return this.getTokenStyle(t, o, n);
      } else if (typeof e == "object")
        return e;
    }
  }
  getTokenColorIndex() {
    if (!this.tokenColorIndex) {
      const e = new On();
      this.tokenColors.forEach((t) => {
        e.add(t.settings.foreground), e.add(t.settings.background);
      }), this.semanticTokenRules.forEach((t) => e.add(t.style.foreground)), Oe.getTokenStylingDefaultRules().forEach((t) => {
        const o = t.defaults[this.type];
        o && typeof o == "object" && e.add(o.foreground);
      }), this.customSemanticTokenRules.forEach((t) => e.add(t.style.foreground)), this.tokenColorIndex = e;
    }
    return this.tokenColorIndex;
  }
  get tokenColorMap() {
    return this.getTokenColorIndex().asArray();
  }
  getTokenStyleMetadata(e, t, o, n = !0, i = {}) {
    const { type: c, language: a } = kt(e, o), l = this.getTokenStyle(c, t, a, n, i);
    if (l)
      return {
        foreground: this.getTokenColorIndex().get(l.foreground),
        bold: l.bold,
        underline: l.underline,
        strikethrough: l.strikethrough,
        italic: l.italic
      };
  }
  getTokenStylingRuleScope(e) {
    if (this.customSemanticTokenRules.indexOf(e) !== -1)
      return "setting";
    if (this.semanticTokenRules.indexOf(e) !== -1)
      return "theme";
  }
  getDefault(e) {
    return Rn.resolveDefaultColor(e, this);
  }
  resolveScopes(e, t) {
    this.themeTokenScopeMatchers || (this.themeTokenScopeMatchers = this.themeTokenColors.map(Nt)), this.customTokenScopeMatchers || (this.customTokenScopeMatchers = this.customTokenColors.map(Nt));
    for (const o of e) {
      let h = function(u, f) {
        for (let m = 0; m < u.length; m++) {
          const p = u[m](o);
          if (p >= 0) {
            const b = f[m], C = f[m].settings;
            p >= c && C.foreground && (n = C.foreground, c = p, d = b), p >= a && M(C.fontStyle) && (i = C.fontStyle, a = p, l = b);
          }
        }
      }, n, i, c = -1, a = -1, l, d;
      if (h(this.themeTokenScopeMatchers, this.themeTokenColors), h(this.customTokenScopeMatchers, this.customTokenColors), n !== void 0 || i !== void 0)
        return t && (t.foreground = d, t.bold = t.italic = t.underline = t.strikethrough = l, t.scope = o), Ae.fromSettings(n, i);
    }
  }
  defines(e) {
    return this.customColorMap.hasOwnProperty(e) || this.colorMap.hasOwnProperty(e);
  }
  setCustomizations(e) {
    this.setCustomColors(e.colorCustomizations), this.setCustomTokenColors(e.tokenColorCustomizations), this.setCustomSemanticTokenColors(e.semanticTokenColorCustomizations);
  }
  setCustomColors(e) {
    this.customColorMap = {}, this.overwriteCustomColors(e);
    const t = this.getThemeSpecificColors(e);
    Me(t) && this.overwriteCustomColors(t), this.tokenColorIndex = void 0, this.textMateThemingRules = void 0, this.customTokenScopeMatchers = void 0;
  }
  overwriteCustomColors(e) {
    for (const t in e) {
      const o = e[t];
      typeof o == "string" && (this.customColorMap[t] = V.fromHex(o));
    }
  }
  setCustomTokenColors(e) {
    this.customTokenColors = [], this.customSemanticHighlightingDeprecated = void 0, this.addCustomTokenColors(e);
    const t = this.getThemeSpecificColors(e);
    Me(t) && this.addCustomTokenColors(t), this.tokenColorIndex = void 0, this.textMateThemingRules = void 0, this.customTokenScopeMatchers = void 0;
  }
  setCustomSemanticTokenColors(e) {
    if (this.customSemanticTokenRules = [], this.customSemanticHighlighting = void 0, e) {
      this.customSemanticHighlighting = e.enabled, e.rules && this.readSemanticTokenRules(e.rules);
      const t = this.getThemeSpecificColors(e);
      Me(t) && (t.enabled !== void 0 && (this.customSemanticHighlighting = t.enabled), t.rules && this.readSemanticTokenRules(t.rules));
    }
    this.tokenColorIndex = void 0, this.textMateThemingRules = void 0;
  }
  isThemeScope(e) {
    return e.charAt(0) === Po && e.charAt(e.length - 1) === Mo;
  }
  isThemeScopeMatch(e) {
    const t = e.charAt(0), o = e.charAt(e.length - 1), n = e.slice(0, -1), i = e.slice(1, -1), c = e.slice(1);
    return e === this.settingsId || this.settingsId.includes(i) && t === xe && o === xe || this.settingsId.startsWith(n) && o === xe || this.settingsId.endsWith(c) && t === xe;
  }
  getThemeSpecificColors(e) {
    let t;
    for (const o in e) {
      const n = e[o];
      if (this.isThemeScope(o) && n instanceof Object && !Array.isArray(n)) {
        const i = o.match(Oo) || [];
        for (const c of i) {
          const a = c.substring(1, c.length - 1);
          if (this.isThemeScopeMatch(a)) {
            t || (t = {});
            const l = n;
            for (const d in l) {
              const h = t[d], u = l[d];
              Array.isArray(h) && Array.isArray(u) ? t[d] = h.concat(u) : u && (t[d] = u);
            }
          }
        }
      }
    }
    return t;
  }
  readSemanticTokenRules(e) {
    for (const t in e)
      if (!this.isThemeScope(t))
        try {
          const o = ro(t, e[t]);
          o && this.customSemanticTokenRules.push(o);
        } catch {
        }
  }
  addCustomTokenColors(e) {
    for (const t in Lt) {
      const o = t, n = e[o];
      if (n) {
        const i = typeof n == "string" ? { foreground: n } : n, c = Lt[o];
        for (const a of c)
          this.customTokenColors.push({ scope: a, settings: i });
      }
    }
    if (Array.isArray(e.textMateRules))
      for (const t of e.textMateRules)
        t.scope && t.settings && this.customTokenColors.push(t);
    e.semanticHighlighting !== void 0 && (this.customSemanticHighlightingDeprecated = e.semanticHighlighting);
  }
  ensureLoaded(e) {
    return this.isLoaded ? Promise.resolve(void 0) : this.load(e);
  }
  reload(e) {
    return this.load(e);
  }
  load(e) {
    if (!this.location)
      return Promise.resolve(void 0);
    this.themeTokenColors = [], this.clearCaches();
    const t = {
      colors: {},
      textMateRules: [],
      semanticTokenRules: [],
      semanticHighlighting: !1
    };
    return so(e, this.location, t).then((o) => {
      this.isLoaded = !0, this.semanticTokenRules = t.semanticTokenRules, this.colorMap = t.colors, this.themeTokenColors = t.textMateRules, this.themeSemanticHighlighting = t.semanticHighlighting;
    });
  }
  clearCaches() {
    this.tokenColorIndex = void 0, this.textMateThemingRules = void 0, this.themeTokenScopeMatchers = void 0, this.customTokenScopeMatchers = void 0;
  }
  toStorage(e) {
    const t = {};
    for (const n in this.colorMap)
      t[n] = V.Format.CSS.formatHexA(this.colorMap[n], !0);
    const o = JSON.stringify({
      id: this.id,
      label: this.label,
      settingsId: this.settingsId,
      themeTokenColors: this.themeTokenColors.map((n) => ({ settings: n.settings, scope: n.scope })),
      semanticTokenRules: this.semanticTokenRules.map(vt.toJSONObject),
      extensionData: j.toJSONObject(this.extensionData),
      themeSemanticHighlighting: this.themeSemanticHighlighting,
      colorMap: t,
      watch: this.watch
    });
    e.store($.STORAGE_KEY, o, 0, 0);
  }
  get baseTheme() {
    return this.classNames[0];
  }
  get classNames() {
    return this.id.split(" ");
  }
  get type() {
    switch (this.baseTheme) {
      case ze:
        return w.LIGHT;
      case Be:
        return w.HIGH_CONTRAST_DARK;
      case Ue:
        return w.HIGH_CONTRAST_LIGHT;
      default:
        return w.DARK;
    }
  }
  static createUnloadedThemeForThemeType(e, t) {
    return $.createUnloadedTheme(Lo(e), t);
  }
  static createUnloadedTheme(e, t) {
    const o = new $(e, "", "__" + e);
    if (o.isLoaded = !1, o.themeTokenColors = [], o.watch = !1, t)
      for (const n in t)
        o.colorMap[n] = V.fromHex(t[n]);
    return o;
  }
  static createLoadedEmptyTheme(e, t) {
    const o = new $(e, "", t);
    return o.isLoaded = !0, o.themeTokenColors = [], o.watch = !1, o;
  }
  static fromStorageData(e) {
    const t = e.get($.STORAGE_KEY, 0);
    if (t)
      try {
        const o = JSON.parse(t), n = new $("", "", "");
        for (const i in o)
          switch (i) {
            case "colorMap": {
              const c = o[i];
              for (const a in c)
                n.colorMap[a] = V.fromHex(c[a]);
              break;
            }
            case "themeTokenColors":
            case "id":
            case "label":
            case "settingsId":
            case "watch":
            case "themeSemanticHighlighting":
              n[i] = o[i];
              break;
            case "semanticTokenRules": {
              const c = o[i];
              if (Array.isArray(c))
                for (const a of c) {
                  const l = vt.fromJSONObject(Oe, a);
                  l && n.semanticTokenRules.push(l);
                }
              break;
            }
            case "location":
              break;
            case "extensionData":
              n.extensionData = j.fromJSONObject(o.extensionData);
              break;
          }
        return !n.id || !n.settingsId ? void 0 : n;
      } catch {
        return;
      }
  }
  static fromExtensionTheme(e, t, o) {
    const n = e.uiTheme || "vs-dark", i = _n(o.extensionId, e.path), c = `${n} ${i}`, a = e.label || at(e.path), l = e.id || a, d = new $(c, a, l);
    return d.description = e.description, d.watch = e._watch === !0, d.location = t, d.extensionData = o, d.isLoaded = !1, d;
  }
};
$.STORAGE_KEY = "colorThemeData";
let z = $;
function _n(s, e) {
  e.startsWith("./") && (e = e.substr(2));
  let t = `${s}-${e}`;
  return t = t.replace(/[^_a-zA-Z0-9-]/g, "-"), t.charAt(0).match(/[0-9-]/) && (t = "_" + t), t;
}
async function so(s, e, t) {
  if (Ho(e) === ".json") {
    const o = await s.readExtensionResource(e), n = [], i = ht(o, n);
    if (n.length > 0)
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/common/colorThemeData",
        "error.cannotparsejson",
        "Problems parsing JSON theme file: {0}",
        n.map((d) => lt(d.error)).join(", ")
      )));
    if (mt(i) !== "object")
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/common/colorThemeData",
        "error.invalidformat",
        "Invalid format for JSON theme file: Object expected."
      )));
    if (i.include && await so(s, X(Fe(e), i.include), t), Array.isArray(i.settings))
      return io(i.settings, t), null;
    t.semanticHighlighting = t.semanticHighlighting || i.semanticHighlighting;
    const c = i.colors;
    if (c) {
      if (typeof c != "object")
        return Promise.reject(new Error(r(
          "vs/workbench/services/themes/common/colorThemeData",
          { key: "error.invalidformat.colors", comment: ["{0} will be replaced by a path. Values in quotes should not be translated."] },
          "Problem parsing color theme file: {0}. Property 'colors' is not of type 'object'.",
          e.toString()
        )));
      for (const d in c)
        typeof c[d] == "string" && (t.colors[d] = V.fromHex(c[d]));
    }
    const a = i.tokenColors;
    if (a)
      if (Array.isArray(a))
        t.textMateRules.push(...a);
      else if (typeof a == "string")
        await Ht(s, X(Fe(e), a), t);
      else
        return Promise.reject(new Error(r(
          "vs/workbench/services/themes/common/colorThemeData",
          { key: "error.invalidformat.tokenColors", comment: ["{0} will be replaced by a path. Values in quotes should not be translated."] },
          "Problem parsing color theme file: {0}. Property 'tokenColors' should be either an array specifying colors or a path to a TextMate theme file",
          e.toString()
        )));
    const l = i.semanticTokenColors;
    if (l && typeof l == "object")
      for (const d in l)
        try {
          const h = ro(d, l[d]);
          h && t.semanticTokenRules.push(h);
        } catch {
          return Promise.reject(new Error(r(
            "vs/workbench/services/themes/common/colorThemeData",
            { key: "error.invalidformat.semanticTokenColors", comment: ["{0} will be replaced by a path. Values in quotes should not be translated."] },
            "Problem parsing color theme file: {0}. Property 'semanticTokenColors' contains a invalid selector",
            e.toString()
          )));
        }
  } else
    return Ht(s, e, t);
}
function Ht(s, e, t) {
  return s.readExtensionResource(e).then((o) => {
    try {
      const i = vn(o).settings;
      return Array.isArray(i) ? (io(i, t), Promise.resolve(null)) : Promise.reject(new Error(r(
        "vs/workbench/services/themes/common/colorThemeData",
        "error.plist.invalidformat",
        "Problem parsing tmTheme file: {0}. 'settings' is not array."
      )));
    } catch (n) {
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/common/colorThemeData",
        "error.cannotparse",
        "Problems parsing tmTheme file: {0}",
        n.message
      )));
    }
  }, (o) => Promise.reject(new Error(r(
    "vs/workbench/services/themes/common/colorThemeData",
    "error.cannotload",
    "Problems loading tmTheme file {0}: {1}",
    e.toString(),
    o.message
  ))));
}
const Dn = {
  light: [
    { scope: "token.info-token", settings: { foreground: "#316bcd" } },
    { scope: "token.warn-token", settings: { foreground: "#cd9731" } },
    { scope: "token.error-token", settings: { foreground: "#cd3131" } },
    { scope: "token.debug-token", settings: { foreground: "#800080" } }
  ],
  dark: [
    { scope: "token.info-token", settings: { foreground: "#6796e6" } },
    { scope: "token.warn-token", settings: { foreground: "#cd9731" } },
    { scope: "token.error-token", settings: { foreground: "#f44747" } },
    { scope: "token.debug-token", settings: { foreground: "#b267e6" } }
  ],
  hcLight: [
    { scope: "token.info-token", settings: { foreground: "#316bcd" } },
    { scope: "token.warn-token", settings: { foreground: "#cd9731" } },
    { scope: "token.error-token", settings: { foreground: "#cd3131" } },
    { scope: "token.debug-token", settings: { foreground: "#800080" } }
  ],
  hcDark: [
    { scope: "token.info-token", settings: { foreground: "#6796e6" } },
    { scope: "token.warn-token", settings: { foreground: "#008000" } },
    { scope: "token.error-token", settings: { foreground: "#FF0000" } },
    { scope: "token.debug-token", settings: { foreground: "#b267e6" } }
  ]
}, At = (s) => -1;
function Ft(s, e) {
  function t(i, c) {
    for (let a = c - 1; a >= 0; a--)
      if (Pn(i, s[a]))
        return a;
    return -1;
  }
  if (e.length < s.length)
    return -1;
  let o = e.length - 1, n = t(e[o--], s.length);
  if (n >= 0) {
    const i = (n + 1) * 65536 + s[n].length;
    for (; o >= 0; )
      if (n = t(e[o--], n), n === -1)
        return -1;
    return i;
  }
  return -1;
}
function Pn(s, e) {
  if (!s)
    return !1;
  if (s === e)
    return !0;
  const t = e.length;
  return s.length > t && s.substr(0, t) === e && s[t] === ".";
}
function Nt(s) {
  const e = s.scope;
  if (!e || !s.settings)
    return At;
  const t = [];
  if (Array.isArray(e))
    for (const o of e)
      Mt(o, Ft, t);
  else
    Mt(e, Ft, t);
  return t.length === 0 ? At : (o) => {
    let n = t[0].matcher(o);
    for (let i = 1; i < t.length; i++)
      n = Math.max(n, t[i].matcher(o));
    return n;
  };
}
function ro(s, e) {
  const t = Oe.parseTokenSelector(s);
  let o;
  if (typeof e == "string" ? o = Ae.fromSettings(e, void 0) : Mn(e) && (o = Ae.fromSettings(e.foreground, e.fontStyle, e.bold, e.underline, e.strikethrough, e.italic)), o)
    return { selector: t, style: o };
}
function Mn(s) {
  return s && (M(s.foreground) || M(s.fontStyle) || Re(s.italic) || Re(s.underline) || Re(s.strikethrough) || Re(s.bold));
}
class On {
  constructor() {
    this._lastColorId = 0, this._id2color = [], this._color2id = /* @__PURE__ */ Object.create(null);
  }
  add(e) {
    if (e = re(e), e === void 0)
      return 0;
    let t = this._color2id[e];
    return t || (t = ++this._lastColorId, this._color2id[e] = t, this._id2color[t] = e, t);
  }
  get(e) {
    if (e = re(e), e === void 0)
      return 0;
    const t = this._color2id[e];
    return t || (console.log(`Color ${e} not in index.`), 0);
  }
  asArray() {
    return this._id2color.slice(0);
  }
}
function re(s) {
  if (!s)
    return;
  typeof s != "string" && (s = V.Format.CSS.formatHexA(s, !0));
  const e = s.length;
  if (s.charCodeAt(0) !== 35 || e !== 4 && e !== 5 && e !== 7 && e !== 9)
    return;
  const t = [35];
  for (let o = 1; o < e; o++) {
    const n = Ln(s.charCodeAt(o));
    if (!n)
      return;
    t.push(n), (e === 4 || e === 5) && t.push(n);
  }
  return t.length === 9 && t[7] === 70 && t[8] === 70 && (t.length = 7), String.fromCharCode(...t);
}
function Ln(s) {
  return s >= 48 && s <= 57 || s >= 65 && s <= 70 ? s : s >= 97 && s <= 102 ? s - 97 + 65 : 0;
}
const St = "^([\\w_-]+)$", Ct = "^(normal|italic|(oblique[ \\w\\s-]+))$", It = "^(normal|bold|lighter|bolder|(\\d{0-1000}))$", $t = "^([\\w .%_-]+)$", Hn = "^woff|woff2|truetype|opentype|embedded-opentype|svg$", An = "vscode://schemas/product-icon-theme", Fn = {
  type: "object",
  allowComments: !0,
  allowTrailingCommas: !0,
  properties: {
    fonts: {
      type: "array",
      items: {
        type: "object",
        properties: {
          id: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/productIconThemeSchema",
              "schema.id",
              "The ID of the font."
            ),
            pattern: St,
            patternErrorMessage: r(
              "vs/workbench/services/themes/common/productIconThemeSchema",
              "schema.id.formatError",
              "The ID must only contain letters, numbers, underscore and minus."
            )
          },
          src: {
            type: "array",
            description: r(
              "vs/workbench/services/themes/common/productIconThemeSchema",
              "schema.src",
              "The location of the font."
            ),
            items: {
              type: "object",
              properties: {
                path: {
                  type: "string",
                  description: r(
                    "vs/workbench/services/themes/common/productIconThemeSchema",
                    "schema.font-path",
                    "The font path, relative to the current product icon theme file."
                  )
                },
                format: {
                  type: "string",
                  description: r(
                    "vs/workbench/services/themes/common/productIconThemeSchema",
                    "schema.font-format",
                    "The format of the font."
                  ),
                  enum: ["woff", "woff2", "truetype", "opentype", "embedded-opentype", "svg"]
                }
              },
              required: [
                "path",
                "format"
              ]
            }
          },
          weight: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/productIconThemeSchema",
              "schema.font-weight",
              "The weight of the font. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight for valid values."
            ),
            anyOf: [
              { enum: ["normal", "bold", "lighter", "bolder"] },
              { type: "string", pattern: It }
            ]
          },
          style: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/productIconThemeSchema",
              "schema.font-style",
              "The style of the font. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-style for valid values."
            ),
            anyOf: [
              { enum: ["normal", "italic", "oblique"] },
              { type: "string", pattern: Ct }
            ]
          }
        },
        required: [
          "id",
          "src"
        ]
      }
    },
    iconDefinitions: {
      description: r(
        "vs/workbench/services/themes/common/productIconThemeSchema",
        "schema.iconDefinitions",
        "Association of icon name to a font character."
      ),
      $ref: Fo
    }
  }
};
function Nn() {
  J.as(dt.JSONContribution).registerSchema(An, Fn);
}
const $n = "vscode://schemas/icon-theme", jn = {
  type: "object",
  allowComments: !0,
  allowTrailingCommas: !0,
  definitions: {
    folderExpanded: {
      type: "string",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.folderExpanded",
        "The folder icon for expanded folders. The expanded folder icon is optional. If not set, the icon defined for folder will be shown."
      )
    },
    folder: {
      type: "string",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.folder",
        "The folder icon for collapsed folders, and if folderExpanded is not set, also for expanded folders."
      )
    },
    file: {
      type: "string",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.file",
        "The default file icon, shown for all files that don't match any extension, filename or language id."
      )
    },
    rootFolder: {
      type: "string",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.rootFolder",
        "The folder icon for collapsed root folders, and if rootFolderExpanded is not set, also for expanded root folders."
      )
    },
    rootFolderExpanded: {
      type: "string",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.rootFolderExpanded",
        "The folder icon for expanded root folders. The expanded root folder icon is optional. If not set, the icon defined for root folder will be shown."
      )
    },
    rootFolderNames: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.rootFolderNames",
        "Associates root folder names to icons. The object key is the root folder name. No patterns or wildcards are allowed. Root folder name matching is case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.folderName",
          "The ID of the icon definition for the association."
        )
      }
    },
    rootFolderNamesExpanded: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.rootFolderNamesExpanded",
        "Associates root folder names to icons for expanded root folders. The object key is the root folder name. No patterns or wildcards are allowed. Root folder name matching is case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.rootFolderNameExpanded",
          "The ID of the icon definition for the association."
        )
      }
    },
    folderNames: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.folderNames",
        "Associates folder names to icons. The object key is the folder name, not including any path segments. No patterns or wildcards are allowed. Folder name matching is case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.folderName",
          "The ID of the icon definition for the association."
        )
      }
    },
    folderNamesExpanded: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.folderNamesExpanded",
        "Associates folder names to icons for expanded folders. The object key is the folder name, not including any path segments. No patterns or wildcards are allowed. Folder name matching is case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.folderNameExpanded",
          "The ID of the icon definition for the association."
        )
      }
    },
    fileExtensions: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.fileExtensions",
        "Associates file extensions to icons. The object key is the file extension name. The extension name is the last segment of a file name after the last dot (not including the dot). Extensions are compared case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.fileExtension",
          "The ID of the icon definition for the association."
        )
      }
    },
    fileNames: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.fileNames",
        "Associates file names to icons. The object key is the full file name, but not including any path segments. File name can include dots and a possible file extension. No patterns or wildcards are allowed. File name matching is case insensitive."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.fileName",
          "The ID of the icon definition for the association."
        )
      }
    },
    languageIds: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.languageIds",
        "Associates languages to icons. The object key is the language id as defined in the language contribution point."
      ),
      additionalProperties: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.languageId",
          "The ID of the icon definition for the association."
        )
      }
    },
    associations: {
      type: "object",
      properties: {
        folderExpanded: {
          $ref: "#/definitions/folderExpanded"
        },
        folder: {
          $ref: "#/definitions/folder"
        },
        file: {
          $ref: "#/definitions/file"
        },
        folderNames: {
          $ref: "#/definitions/folderNames"
        },
        folderNamesExpanded: {
          $ref: "#/definitions/folderNamesExpanded"
        },
        rootFolder: {
          $ref: "#/definitions/rootFolder"
        },
        rootFolderExpanded: {
          $ref: "#/definitions/rootFolderExpanded"
        },
        rootFolderNames: {
          $ref: "#/definitions/rootFolderNames"
        },
        rootFolderNamesExpanded: {
          $ref: "#/definitions/rootFolderNamesExpanded"
        },
        fileExtensions: {
          $ref: "#/definitions/fileExtensions"
        },
        fileNames: {
          $ref: "#/definitions/fileNames"
        },
        languageIds: {
          $ref: "#/definitions/languageIds"
        }
      }
    }
  },
  properties: {
    fonts: {
      type: "array",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.fonts",
        "Fonts that are used in the icon definitions."
      ),
      items: {
        type: "object",
        properties: {
          id: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.id",
              "The ID of the font."
            ),
            pattern: St,
            patternErrorMessage: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.id.formatError",
              "The ID must only contain letter, numbers, underscore and minus."
            )
          },
          src: {
            type: "array",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.src",
              "The location of the font."
            ),
            items: {
              type: "object",
              properties: {
                path: {
                  type: "string",
                  description: r(
                    "vs/workbench/services/themes/common/fileIconThemeSchema",
                    "schema.font-path",
                    "The font path, relative to the current file icon theme file."
                  )
                },
                format: {
                  type: "string",
                  description: r(
                    "vs/workbench/services/themes/common/fileIconThemeSchema",
                    "schema.font-format",
                    "The format of the font."
                  ),
                  enum: ["woff", "woff2", "truetype", "opentype", "embedded-opentype", "svg"]
                }
              },
              required: [
                "path",
                "format"
              ]
            }
          },
          weight: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.font-weight",
              "The weight of the font. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-weight for valid values."
            ),
            pattern: It
          },
          style: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.font-style",
              "The style of the font. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-style for valid values."
            ),
            pattern: Ct
          },
          size: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.font-size",
              "The default size of the font. See https://developer.mozilla.org/en-US/docs/Web/CSS/font-size for valid values."
            ),
            pattern: $t
          }
        },
        required: [
          "id",
          "src"
        ]
      }
    },
    iconDefinitions: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.iconDefinitions",
        "Description of all icons that can be used when associating files to icons."
      ),
      additionalProperties: {
        type: "object",
        description: r(
          "vs/workbench/services/themes/common/fileIconThemeSchema",
          "schema.iconDefinition",
          "An icon definition. The object key is the ID of the definition."
        ),
        properties: {
          iconPath: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.iconPath",
              "When using a SVG or PNG: The path to the image. The path is relative to the icon set file."
            )
          },
          fontCharacter: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.fontCharacter",
              "When using a glyph font: The character in the font to use."
            )
          },
          fontColor: {
            type: "string",
            format: "color-hex",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.fontColor",
              "When using a glyph font: The color to use."
            )
          },
          fontSize: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.fontSize",
              "When using a font: The font size in percentage to the text font. If not set, defaults to the size in the font definition."
            ),
            pattern: $t
          },
          fontId: {
            type: "string",
            description: r(
              "vs/workbench/services/themes/common/fileIconThemeSchema",
              "schema.fontId",
              "When using a font: The id of the font. If not set, defaults to the first font definition."
            )
          }
        }
      }
    },
    folderExpanded: {
      $ref: "#/definitions/folderExpanded"
    },
    folder: {
      $ref: "#/definitions/folder"
    },
    file: {
      $ref: "#/definitions/file"
    },
    folderNames: {
      $ref: "#/definitions/folderNames"
    },
    folderNamesExpanded: {
      $ref: "#/definitions/folderNamesExpanded"
    },
    rootFolder: {
      $ref: "#/definitions/rootFolder"
    },
    rootFolderExpanded: {
      $ref: "#/definitions/rootFolderExpanded"
    },
    rootFolderNames: {
      $ref: "#/definitions/rootFolderNames"
    },
    rootFolderNamesExpanded: {
      $ref: "#/definitions/rootFolderNamesExpanded"
    },
    fileExtensions: {
      $ref: "#/definitions/fileExtensions"
    },
    fileNames: {
      $ref: "#/definitions/fileNames"
    },
    languageIds: {
      $ref: "#/definitions/languageIds"
    },
    light: {
      $ref: "#/definitions/associations",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.light",
        "Optional associations for file icons in light color themes."
      )
    },
    highContrast: {
      $ref: "#/definitions/associations",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.highContrast",
        "Optional associations for file icons in high contrast color themes."
      )
    },
    hidesExplorerArrows: {
      type: "boolean",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.hidesExplorerArrows",
        "Configures whether the file explorer's arrows should be hidden when this theme is active."
      )
    },
    showLanguageModeIcons: {
      type: "boolean",
      description: r(
        "vs/workbench/services/themes/common/fileIconThemeSchema",
        "schema.showLanguageModeIcons",
        "Configures whether the default language icons should be used if the theme does not define an icon for a language."
      )
    }
  }
};
function Gn() {
  J.as(dt.JSONContribution).registerSchema($n, jn);
}
const O = class O {
  constructor(e, t, o) {
    this.id = e, this.label = t, this.settingsId = o, this.isLoaded = !1, this.hasFileIcons = !1, this.hasFolderIcons = !1, this.hidesExplorerArrows = !1;
  }
  ensureLoaded(e) {
    return this.isLoaded ? Promise.resolve(this.styleSheetContent) : this.load(e);
  }
  reload(e) {
    return this.load(e);
  }
  load(e) {
    return e.load(this);
  }
  static fromExtensionTheme(e, t, o) {
    const n = o.extensionId + "-" + e.id, i = e.label || at(e.path), c = e.id, a = new O(n, i, c);
    return a.description = e.description, a.location = t, a.extensionData = o, a.watch = e._watch, a.isLoaded = !1, a;
  }
  static get noIconTheme() {
    let e = O._noIconTheme;
    return e || (e = O._noIconTheme = new O("", "", null), e.hasFileIcons = !1, e.hasFolderIcons = !1, e.hidesExplorerArrows = !1, e.isLoaded = !0, e.extensionData = void 0, e.watch = !1), e;
  }
  static createUnloadedTheme(e) {
    const t = new O(e, "", "__" + e);
    return t.isLoaded = !1, t.hasFileIcons = !1, t.hasFolderIcons = !1, t.hidesExplorerArrows = !1, t.extensionData = void 0, t.watch = !1, t;
  }
  static fromStorageData(e) {
    const t = e.get(O.STORAGE_KEY, 0);
    if (t)
      try {
        const o = JSON.parse(t), n = new O("", "", null);
        for (const i in o)
          switch (i) {
            case "id":
            case "label":
            case "description":
            case "settingsId":
            case "styleSheetContent":
            case "hasFileIcons":
            case "hidesExplorerArrows":
            case "hasFolderIcons":
            case "watch":
              n[i] = o[i];
              break;
            case "location":
              break;
            case "extensionData":
              n.extensionData = j.fromJSONObject(o.extensionData);
              break;
          }
        return n;
      } catch {
        return;
      }
  }
  toStorage(e) {
    const t = JSON.stringify({
      id: this.id,
      label: this.label,
      description: this.description,
      settingsId: this.settingsId,
      styleSheetContent: this.styleSheetContent,
      hasFileIcons: this.hasFileIcons,
      hasFolderIcons: this.hasFolderIcons,
      hidesExplorerArrows: this.hidesExplorerArrows,
      extensionData: j.toJSONObject(this.extensionData),
      watch: this.watch
    });
    e.store(O.STORAGE_KEY, t, 0, 1);
  }
};
O.STORAGE_KEY = "iconThemeData", O._noIconTheme = null;
let q = O;
class Un {
  constructor(e, t) {
    this.fileService = e, this.languageService = t;
  }
  load(e) {
    return e.location ? this.loadIconThemeDocument(e.location).then((t) => {
      const o = this.processIconThemeDocument(e.id, e.location, t);
      return e.styleSheetContent = o.content, e.hasFileIcons = o.hasFileIcons, e.hasFolderIcons = o.hasFolderIcons, e.hidesExplorerArrows = o.hidesExplorerArrows, e.isLoaded = !0, e.styleSheetContent;
    }) : Promise.resolve(e.styleSheetContent);
  }
  loadIconThemeDocument(e) {
    return this.fileService.readExtensionResource(e).then((t) => {
      const o = [], n = ht(t, o);
      return o.length > 0 ? Promise.reject(new Error(r(
        "vs/workbench/services/themes/browser/fileIconThemeData",
        "error.cannotparseicontheme",
        "Problems parsing file icons file: {0}",
        o.map((i) => lt(i.error)).join(", ")
      ))) : mt(n) !== "object" ? Promise.reject(new Error(r(
        "vs/workbench/services/themes/browser/fileIconThemeData",
        "error.invalidformat",
        "Invalid format for file icons theme file: Object expected."
      ))) : Promise.resolve(n);
    });
  }
  processIconThemeDocument(e, t, o) {
    const n = { content: "", hasFileIcons: !1, hasFolderIcons: !1, hidesExplorerArrows: !!o.hidesExplorerArrows };
    let i = !1;
    if (!o.iconDefinitions)
      return n;
    const c = {}, a = {}, l = Fe(t);
    function d(b) {
      return X(l, b);
    }
    function h(b, C) {
      function T(I, P) {
        if (P) {
          let y = c[P];
          y || (y = c[P] = []), y.push(I);
        }
      }
      if (b) {
        let I = ".show-file-icons";
        C && (I = C + " " + I);
        const P = ".monaco-tl-twistie.collapsible:not(.collapsed) + .monaco-tl-contents";
        b.folder && (T(`${I} .folder-icon::before`, b.folder), n.hasFolderIcons = !0), b.folderExpanded && (T(`${I} ${P} .folder-icon::before`, b.folderExpanded), n.hasFolderIcons = !0);
        const y = b.rootFolder || b.folder, te = b.rootFolderExpanded || b.folderExpanded;
        y && (T(`${I} .rootfolder-icon::before`, y), n.hasFolderIcons = !0), te && (T(`${I} ${P} .rootfolder-icon::before`, te), n.hasFolderIcons = !0), b.file && (T(`${I} .file-icon::before`, b.file), n.hasFileIcons = !0);
        const G = b.folderNames;
        if (G)
          for (const x in G) {
            const E = [], U = De(x.toLowerCase(), E);
            E.push(`.${N(U)}-name-folder-icon`), T(`${I} ${E.join("")}.folder-icon::before`, G[x]), n.hasFolderIcons = !0;
          }
        const Y = b.folderNamesExpanded;
        if (Y)
          for (const x in Y) {
            const E = [], U = De(x.toLowerCase(), E);
            E.push(`.${N(U)}-name-folder-icon`), T(`${I} ${P} ${E.join("")}.folder-icon::before`, Y[x]), n.hasFolderIcons = !0;
          }
        const H = b.rootFolderNames;
        if (H)
          for (const x in H) {
            const E = x.toLowerCase();
            T(`${I} .${N(E)}-root-name-folder-icon.rootfolder-icon::before`, H[x]), n.hasFolderIcons = !0;
          }
        const fe = b.rootFolderNamesExpanded;
        if (fe)
          for (const x in fe) {
            const E = x.toLowerCase();
            T(`${I} ${P} .${N(E)}-root-name-folder-icon.rootfolder-icon::before`, fe[x]), n.hasFolderIcons = !0;
          }
        const A = b.languageIds;
        if (A) {
          !A.jsonc && A.json && (A.jsonc = A.json);
          for (const x in A)
            T(`${I} .${N(x)}-lang-file-icon.file-icon::before`, A[x]), n.hasFileIcons = !0, i = !0, a[x] = !0;
        }
        const ge = b.fileExtensions;
        if (ge)
          for (const x in ge) {
            const E = [], B = De(x.toLowerCase(), E).split(".");
            if (B.length) {
              for (let F = 0; F < B.length; F++)
                E.push(`.${N(B.slice(F).join("."))}-ext-file-icon`);
              E.push(".ext-file-icon");
            }
            T(`${I} ${E.join("")}.file-icon::before`, ge[x]), n.hasFileIcons = !0, i = !0;
          }
        const pe = b.fileNames;
        if (pe)
          for (const x in pe) {
            const E = [], U = De(x.toLowerCase(), E);
            E.push(`.${N(U)}-name-file-icon`), E.push(".name-file-icon");
            const B = U.split(".");
            if (B.length) {
              for (let F = 1; F < B.length; F++)
                E.push(`.${N(B.slice(F).join("."))}-ext-file-icon`);
              E.push(".ext-file-icon");
            }
            T(`${I} ${E.join("")}.file-icon::before`, pe[x]), n.hasFileIcons = !0, i = !0;
          }
      }
    }
    if (h(o), h(o.light, ".vs"), h(o.highContrast, ".hc-black"), h(o.highContrast, ".hc-light"), !n.hasFileIcons && !n.hasFolderIcons)
      return n;
    const u = o.showLanguageModeIcons === !0 || i && o.showLanguageModeIcons !== !1, f = [], m = o.fonts, p = /* @__PURE__ */ new Map();
    if (Array.isArray(m)) {
      const b = m[0].size || "150%";
      m.forEach((C) => {
        const T = C.src.map((I) => `${_e(d(I.path))} format('${I.format}')`).join(", ");
        f.push(`@font-face { src: ${T}; font-family: '${C.id}'; font-weight: ${C.weight}; font-style: ${C.style}; font-display: block; }`), C.size !== void 0 && C.size !== b && p.set(C.id, C.size);
      }), f.push(`.show-file-icons .file-icon::before, .show-file-icons .folder-icon::before, .show-file-icons .rootfolder-icon::before { font-family: '${m[0].id}'; font-size: ${b}; }`);
    }
    for (const b in c) {
      const C = c[b], T = o.iconDefinitions[b];
      if (T) {
        if (T.iconPath)
          f.push(`${C.join(", ")} { content: ' '; background-image: ${_e(d(T.iconPath))}; }`);
        else if (T.fontCharacter || T.fontColor) {
          const I = [];
          T.fontColor && I.push(`color: ${T.fontColor};`), T.fontCharacter && I.push(`content: '${T.fontCharacter}';`);
          const P = T.fontSize ?? (T.fontId ? p.get(T.fontId) : void 0);
          P && I.push(`font-size: ${P};`), T.fontId && I.push(`font-family: ${T.fontId};`), u && I.push("background-image: unset;"), f.push(`${C.join(", ")} { ${I.join(" ")} }`);
        }
      }
    }
    if (u) {
      for (const b of this.languageService.getRegisteredLanguageIds())
        if (!a[b]) {
          const C = this.languageService.getIcon(b);
          if (C) {
            const T = `.show-file-icons .${N(b)}-lang-file-icon.file-icon::before`;
            f.push(`${T} { content: ' '; background-image: ${_e(C.dark)}; }`), f.push(`.vs ${T} { content: ' '; background-image: ${_e(C.light)}; }`);
          }
        }
    }
    return n.content = f.join(`
`), n;
  }
}
function De(s, e) {
  const t = s.lastIndexOf("/");
  if (t >= 0) {
    const o = s.substring(0, t);
    return e.push(`.${N(o)}-name-dir-icon`), s.substring(t + 1);
  }
  return s;
}
function N(s) {
  return s = s.replace(/[\11\12\14\15\40]/g, "/"), qe.CSS.escape(s);
}
const jt = [
  "comment",
  "comment.block",
  "comment.block.documentation",
  "comment.line",
  "constant",
  "constant.character",
  "constant.character.escape",
  "constant.numeric",
  "constant.numeric.integer",
  "constant.numeric.float",
  "constant.numeric.hex",
  "constant.numeric.octal",
  "constant.other",
  "constant.regexp",
  "constant.rgb-value",
  "emphasis",
  "entity",
  "entity.name",
  "entity.name.class",
  "entity.name.function",
  "entity.name.method",
  "entity.name.section",
  "entity.name.selector",
  "entity.name.tag",
  "entity.name.type",
  "entity.other",
  "entity.other.attribute-name",
  "entity.other.inherited-class",
  "invalid",
  "invalid.deprecated",
  "invalid.illegal",
  "keyword",
  "keyword.control",
  "keyword.operator",
  "keyword.operator.new",
  "keyword.operator.assignment",
  "keyword.operator.arithmetic",
  "keyword.operator.logical",
  "keyword.other",
  "markup",
  "markup.bold",
  "markup.changed",
  "markup.deleted",
  "markup.heading",
  "markup.inline.raw",
  "markup.inserted",
  "markup.italic",
  "markup.list",
  "markup.list.numbered",
  "markup.list.unnumbered",
  "markup.other",
  "markup.quote",
  "markup.raw",
  "markup.underline",
  "markup.underline.link",
  "meta",
  "meta.block",
  "meta.cast",
  "meta.class",
  "meta.function",
  "meta.function-call",
  "meta.preprocessor",
  "meta.return-type",
  "meta.selector",
  "meta.tag",
  "meta.type.annotation",
  "meta.type",
  "punctuation.definition.string.begin",
  "punctuation.definition.string.end",
  "punctuation.separator",
  "punctuation.separator.continuation",
  "punctuation.terminator",
  "storage",
  "storage.modifier",
  "storage.type",
  "string",
  "string.interpolated",
  "string.other",
  "string.quoted",
  "string.quoted.double",
  "string.quoted.other",
  "string.quoted.single",
  "string.quoted.triple",
  "string.regexp",
  "string.unquoted",
  "strong",
  "support",
  "support.class",
  "support.constant",
  "support.function",
  "support.other",
  "support.type",
  "support.type.property-name",
  "support.variable",
  "variable",
  "variable.language",
  "variable.name",
  "variable.other",
  "variable.other.readwrite",
  "variable.parameter"
], Ve = "vscode://schemas/textmate-colors", Bn = `${Ve}#/definitions/colorGroup`, zn = {
  type: "array",
  definitions: {
    colorGroup: {
      default: "#FF0000",
      anyOf: [
        {
          type: "string",
          format: "color-hex"
        },
        {
          $ref: "#/definitions/settings"
        }
      ]
    },
    settings: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/colorThemeSchema",
        "schema.token.settings",
        "Colors and styles for the token."
      ),
      properties: {
        foreground: {
          type: "string",
          description: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.token.foreground",
            "Foreground color for the token."
          ),
          format: "color-hex",
          default: "#ff0000"
        },
        background: {
          type: "string",
          deprecationMessage: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.token.background.warning",
            "Token background colors are currently not supported."
          )
        },
        fontStyle: {
          type: "string",
          description: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.token.fontStyle",
            "Font style of the rule: 'italic', 'bold', 'underline', 'strikethrough' or a combination. The empty string unsets inherited settings."
          ),
          pattern: "^(\\s*\\b(italic|bold|underline|strikethrough))*\\s*$",
          patternErrorMessage: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.fontStyle.error",
            "Font style must be 'italic', 'bold', 'underline', 'strikethrough' or a combination or the empty string."
          ),
          defaultSnippets: [
            { label: r(
              "vs/workbench/services/themes/common/colorThemeSchema",
              "schema.token.fontStyle.none",
              "None (clear inherited style)"
            ), bodyText: '""' },
            { body: "italic" },
            { body: "bold" },
            { body: "underline" },
            { body: "strikethrough" },
            { body: "italic bold" },
            { body: "italic underline" },
            { body: "italic strikethrough" },
            { body: "bold underline" },
            { body: "bold strikethrough" },
            { body: "underline strikethrough" },
            { body: "italic bold underline" },
            { body: "italic bold strikethrough" },
            { body: "italic underline strikethrough" },
            { body: "bold underline strikethrough" },
            { body: "italic bold underline strikethrough" }
          ]
        }
      },
      additionalProperties: !1,
      defaultSnippets: [{ body: { foreground: "${1:#FF0000}", fontStyle: "${2:bold}" } }]
    }
  },
  items: {
    type: "object",
    defaultSnippets: [{ body: { scope: "${1:keyword.operator}", settings: { foreground: "${2:#FF0000}" } } }],
    properties: {
      name: {
        type: "string",
        description: r(
          "vs/workbench/services/themes/common/colorThemeSchema",
          "schema.properties.name",
          "Description of the rule."
        )
      },
      scope: {
        description: r(
          "vs/workbench/services/themes/common/colorThemeSchema",
          "schema.properties.scope",
          "Scope selector against which this rule matches."
        ),
        anyOf: [
          {
            enum: jt
          },
          {
            type: "string"
          },
          {
            type: "array",
            items: {
              enum: jt
            }
          },
          {
            type: "array",
            items: {
              type: "string"
            }
          }
        ]
      },
      settings: {
        $ref: "#/definitions/settings"
      }
    },
    required: [
      "settings"
    ],
    additionalProperties: !1
  }
}, co = "vscode://schemas/color-theme", qn = {
  type: "object",
  allowComments: !0,
  allowTrailingCommas: !0,
  properties: {
    colors: {
      description: r(
        "vs/workbench/services/themes/common/colorThemeSchema",
        "schema.workbenchColors",
        "Colors in the workbench"
      ),
      $ref: ut,
      additionalProperties: !1
    },
    tokenColors: {
      anyOf: [
        {
          type: "string",
          description: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.tokenColors.path",
            "Path to a tmTheme file (relative to the current file)."
          )
        },
        {
          description: r(
            "vs/workbench/services/themes/common/colorThemeSchema",
            "schema.colors",
            "Colors for syntax highlighting"
          ),
          $ref: Ve
        }
      ]
    },
    semanticHighlighting: {
      type: "boolean",
      description: r(
        "vs/workbench/services/themes/common/colorThemeSchema",
        "schema.supportsSemanticHighlighting",
        "Whether semantic highlighting should be enabled for this theme."
      )
    },
    semanticTokenColors: {
      type: "object",
      description: r(
        "vs/workbench/services/themes/common/colorThemeSchema",
        "schema.semanticTokenColors",
        "Colors for semantic tokens"
      ),
      $ref: Jt
    }
  }
};
function Kn() {
  const s = J.as(dt.JSONContribution);
  s.registerSchema(co, qn), s.registerSchema(Ve, zn);
}
function Vn() {
  return ft.registerExtensionPoint({
    extensionPoint: "themes",
    jsonSchema: {
      description: r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "vscode.extension.contributes.themes",
        "Contributes textmate color themes."
      ),
      type: "array",
      items: {
        type: "object",
        defaultSnippets: [{ body: { label: "${1:label}", id: "${2:id}", uiTheme: Ne, path: "./themes/${3:id}.tmTheme." } }],
        properties: {
          id: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.themes.id",
              "Id of the color theme as used in the user settings."
            ),
            type: "string"
          },
          label: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.themes.label",
              "Label of the color theme as shown in the UI."
            ),
            type: "string"
          },
          uiTheme: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.themes.uiTheme",
              "Base theme defining the colors around the editor: 'vs' is the light color theme, 'vs-dark' is the dark color theme. 'hc-black' is the dark high contrast theme, 'hc-light' is the light high contrast theme."
            ),
            enum: [ze, Ne, Be, Ue]
          },
          path: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.themes.path",
              "Path of the tmTheme file. The path is relative to the extension folder and is typically './colorthemes/awesome-color-theme.json'."
            ),
            type: "string"
          }
        },
        required: ["path", "uiTheme"]
      }
    }
  });
}
function Wn() {
  return ft.registerExtensionPoint({
    extensionPoint: "iconThemes",
    jsonSchema: {
      description: r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "vscode.extension.contributes.iconThemes",
        "Contributes file icon themes."
      ),
      type: "array",
      items: {
        type: "object",
        defaultSnippets: [{ body: { id: "${1:id}", label: "${2:label}", path: "./fileicons/${3:id}-icon-theme.json" } }],
        properties: {
          id: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.iconThemes.id",
              "Id of the file icon theme as used in the user settings."
            ),
            type: "string"
          },
          label: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.iconThemes.label",
              "Label of the file icon theme as shown in the UI."
            ),
            type: "string"
          },
          path: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.iconThemes.path",
              "Path of the file icon theme definition file. The path is relative to the extension folder and is typically './fileicons/awesome-icon-theme.json'."
            ),
            type: "string"
          }
        },
        required: ["path", "id"]
      }
    }
  });
}
function Jn() {
  return ft.registerExtensionPoint({
    extensionPoint: "productIconThemes",
    jsonSchema: {
      description: r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "vscode.extension.contributes.productIconThemes",
        "Contributes product icon themes."
      ),
      type: "array",
      items: {
        type: "object",
        defaultSnippets: [{ body: { id: "${1:id}", label: "${2:label}", path: "./producticons/${3:id}-product-icon-theme.json" } }],
        properties: {
          id: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.productIconThemes.id",
              "Id of the product icon theme as used in the user settings."
            ),
            type: "string"
          },
          label: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.productIconThemes.label",
              "Label of the product icon theme as shown in the UI."
            ),
            type: "string"
          },
          path: {
            description: r(
              "vs/workbench/services/themes/common/themeExtensionPoints",
              "vscode.extension.contributes.productIconThemes.path",
              "Path of the product icon theme definition file. The path is relative to the extension folder and is typically './producticons/awesome-product-icon-theme.json'."
            ),
            type: "string"
          }
        },
        required: ["path", "id"]
      }
    }
  });
}
class Yn extends $o {
  constructor() {
    super(...arguments), this.type = "markdown";
  }
  shouldRender(e) {
    var t, o, n;
    return !!((t = e.contributes) != null && t.themes) || !!((o = e.contributes) != null && o.iconThemes) || !!((n = e.contributes) != null && n.productIconThemes);
  }
  render(e) {
    var o, n, i;
    const t = new jo();
    if ((o = e.contributes) != null && o.themes) {
      t.appendMarkdown(`### ${r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "color themes",
        "Color Themes"
      )}

`);
      for (const c of e.contributes.themes)
        t.appendMarkdown(`- ${c.label}
`);
    }
    if ((n = e.contributes) != null && n.iconThemes) {
      t.appendMarkdown(`### ${r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "file icon themes",
        "File Icon Themes"
      )}

`);
      for (const c of e.contributes.iconThemes)
        t.appendMarkdown(`- ${c.label}
`);
    }
    if ((i = e.contributes) != null && i.productIconThemes) {
      t.appendMarkdown(`### ${r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "product icon themes",
        "Product Icon Themes"
      )}

`);
      for (const c of e.contributes.productIconThemes)
        t.appendMarkdown(`- ${c.label}
`);
    }
    return {
      data: t,
      dispose: () => {
      }
    };
  }
}
J.as(No.ExtensionFeaturesRegistry).registerExtensionFeature({
  id: "themes",
  label: r(
    "vs/workbench/services/themes/common/themeExtensionPoints",
    "themes",
    "Themes"
  ),
  access: {
    canToggle: !1
  },
  renderer: new Yt(Yn)
});
class Ye {
  constructor(e, t, o = !1, n = void 0) {
    this.themesExtPoint = e, this.create = t, this.idRequired = o, this.builtInTheme = n, this.onDidChangeEmitter = new Ce(), this.onDidChange = this.onDidChangeEmitter.event, this.extensionThemes = [], this.initialize();
  }
  initialize() {
    this.themesExtPoint.setHandler((e, t) => {
      const o = {}, n = [];
      for (const c of this.extensionThemes)
        o[c.id] = c;
      this.extensionThemes.length = 0;
      for (const c of e) {
        const a = j.fromName(c.description.publisher, c.description.name, c.description.isBuiltin);
        this.onThemes(a, c.description.extensionLocation, c.value, this.extensionThemes, c.collector);
      }
      for (const c of this.extensionThemes)
        o[c.id] ? delete o[c.id] : n.push(c);
      const i = Object.values(o);
      this.onDidChangeEmitter.fire({ themes: this.extensionThemes, added: n, removed: i });
    });
  }
  onThemes(e, t, o, n = [], i) {
    return Array.isArray(o) ? (o.forEach((c) => {
      if (!c.path || !M(c.path)) {
        i == null || i.error(r(
          "vs/workbench/services/themes/common/themeExtensionPoints",
          "reqpath",
          "Expected string in `contributes.{0}.path`. Provided value: {1}",
          this.themesExtPoint.name,
          String(c.path)
        ));
        return;
      }
      if (this.idRequired && (!c.id || !M(c.id))) {
        i == null || i.error(r(
          "vs/workbench/services/themes/common/themeExtensionPoints",
          "reqid",
          "Expected string in `contributes.{0}.id`. Provided value: {1}",
          this.themesExtPoint.name,
          String(c.id)
        ));
        return;
      }
      const a = X(t, c.path);
      Et(a, t) || i == null || i.warn(r(
        "vs/workbench/services/themes/common/themeExtensionPoints",
        "invalid.path.1",
        "Expected `contributes.{0}.path` ({1}) to be included inside extension's folder ({2}). This might make the extension non-portable.",
        this.themesExtPoint.name,
        a.path,
        t.path
      ));
      const l = this.create(c, a, e);
      n.push(l);
    }), n) : (i == null || i.error(r(
      "vs/workbench/services/themes/common/themeExtensionPoints",
      "reqarray",
      "Extension point `{0}` must be an array.",
      this.themesExtPoint.name
    )), n);
  }
  findThemeById(e) {
    if (this.builtInTheme && this.builtInTheme.id === e)
      return this.builtInTheme;
    const t = this.getThemes();
    for (const o of t)
      if (o.id === e)
        return o;
  }
  findThemeBySettingsId(e, t) {
    if (this.builtInTheme && this.builtInTheme.settingsId === e)
      return this.builtInTheme;
    const o = this.getThemes();
    let n;
    for (const i of o) {
      if (i.settingsId === e)
        return i;
      i.settingsId === t && (n = i);
    }
    return n;
  }
  findThemeByExtensionLocation(e) {
    return e ? this.getThemes().filter((t) => t.location && Et(t.location, e)) : [];
  }
  getThemes() {
    return this.extensionThemes;
  }
  getMarketplaceThemes(e, t, o) {
    var i;
    const n = (i = e == null ? void 0 : e.contributes) == null ? void 0 : i[this.themesExtPoint.name];
    return Array.isArray(n) ? this.onThemes(o, t, n) : [];
  }
}
const ue = J.as(Go.Configuration), ne = [], ie = [], se = [];
function W(s) {
  return `\`#${s}#\``;
}
const Qn = {
  type: "string",
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorTheme",
    "Specifies the color theme used in the workbench."
  ),
  default: gt ? v.COLOR_THEME_LIGHT : v.COLOR_THEME_DARK,
  enum: ne,
  enumDescriptions: se,
  enumItemLabels: ie,
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorThemeError",
    "Theme is unknown or not installed."
  )
}, Zn = {
  type: "string",
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "preferredDarkColorTheme", comment: ["{0} will become a link to another setting."] },
    "Specifies the preferred color theme for dark OS appearance when {0} is enabled.",
    W(S.DETECT_COLOR_SCHEME)
  ),
  default: v.COLOR_THEME_DARK,
  enum: ne,
  enumDescriptions: se,
  enumItemLabels: ie,
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorThemeError",
    "Theme is unknown or not installed."
  )
}, Xn = {
  type: "string",
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "preferredLightColorTheme", comment: ["{0} will become a link to another setting."] },
    "Specifies the preferred color theme for light OS appearance when {0} is enabled.",
    W(S.DETECT_COLOR_SCHEME)
  ),
  default: v.COLOR_THEME_LIGHT,
  enum: ne,
  enumDescriptions: se,
  enumItemLabels: ie,
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorThemeError",
    "Theme is unknown or not installed."
  )
}, ei = {
  type: "string",
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "preferredHCDarkColorTheme", comment: ["{0} will become a link to another setting."] },
    "Specifies the preferred color theme used in high contrast dark mode when {0} is enabled.",
    W(S.DETECT_HC)
  ),
  default: v.COLOR_THEME_HC_DARK,
  enum: ne,
  enumDescriptions: se,
  enumItemLabels: ie,
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorThemeError",
    "Theme is unknown or not installed."
  )
}, ti = {
  type: "string",
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "preferredHCLightColorTheme", comment: ["{0} will become a link to another setting."] },
    "Specifies the preferred color theme used in high contrast light mode when {0} is enabled.",
    W(S.DETECT_HC)
  ),
  default: v.COLOR_THEME_HC_LIGHT,
  enum: ne,
  enumDescriptions: se,
  enumItemLabels: ie,
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "colorThemeError",
    "Theme is unknown or not installed."
  )
}, oi = {
  type: "boolean",
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "detectColorScheme", comment: ["{0} and {1} will become links to other settings."] },
    "If set, automatically switch to the preferred color theme based on the OS appearance. If the OS appearance is dark, the theme specified at {0} is used, for light {1}.",
    W(S.PREFERRED_DARK_THEME),
    W(S.PREFERRED_LIGHT_THEME)
  ),
  default: !1
}, ao = {
  type: "object",
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "workbenchColors",
    "Overrides colors from the currently selected color theme."
  ),
  allOf: [{ $ref: ut }],
  default: {},
  defaultSnippets: [{
    body: {}
  }]
}, Le = {
  type: ["string", "null"],
  default: v.FILE_ICON_THEME,
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "iconTheme",
    "Specifies the file icon theme used in the workbench or 'null' to not show any file icons."
  ),
  enum: [null],
  enumItemLabels: [r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "noIconThemeLabel",
    "None"
  )],
  enumDescriptions: [r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "noIconThemeDesc",
    "No file icons"
  )],
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "iconThemeError",
    "File icon theme is unknown or not installed."
  )
}, He = {
  type: ["string", "null"],
  default: v.PRODUCT_ICON_THEME,
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "productIconTheme",
    "Specifies the product icon theme used."
  ),
  enum: [v.PRODUCT_ICON_THEME],
  enumItemLabels: [r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "defaultProductIconThemeLabel",
    "Default"
  )],
  enumDescriptions: [r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "defaultProductIconThemeDesc",
    "Default"
  )],
  errorMessage: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "productIconThemeError",
    "Product icon theme is unknown or not installed."
  )
}, ni = {
  type: "boolean",
  default: !0,
  markdownDescription: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    { key: "autoDetectHighContrast", comment: ["{0} and {1} will become links to other settings."] },
    "If enabled, will automatically change to high contrast theme if the OS is using a high contrast theme. The high contrast theme to use is specified by {0} and {1}.",
    W(S.PREFERRED_HC_DARK_THEME),
    W(S.PREFERRED_HC_LIGHT_THEME)
  ),
  scope: 1
}, We = {
  id: "workbench",
  order: 7.1,
  type: "object",
  properties: {
    [S.COLOR_THEME]: Qn,
    [S.PREFERRED_DARK_THEME]: Zn,
    [S.PREFERRED_LIGHT_THEME]: Xn,
    [S.PREFERRED_HC_DARK_THEME]: ei,
    [S.PREFERRED_HC_LIGHT_THEME]: ti,
    [S.FILE_ICON_THEME]: Le,
    [S.COLOR_CUSTOMIZATIONS]: ao,
    [S.PRODUCT_ICON_THEME]: He
  }
};
ue.registerConfiguration(We);
const ii = {
  id: "window",
  order: 8.1,
  type: "object",
  properties: {
    [S.DETECT_HC]: ni,
    [S.DETECT_COLOR_SCHEME]: oi
  }
};
ue.registerConfiguration(ii);
function oe(s) {
  return {
    description: s,
    $ref: Bn
  };
}
const Qe = "^\\[[^\\]]*(\\]\\s*\\[[^\\]]*)*\\]$", ho = {
  type: "object",
  properties: {
    comments: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.comments",
      "Sets the colors and styles for comments"
    )),
    strings: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.strings",
      "Sets the colors and styles for strings literals."
    )),
    keywords: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.keywords",
      "Sets the colors and styles for keywords."
    )),
    numbers: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.numbers",
      "Sets the colors and styles for number literals."
    )),
    types: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.types",
      "Sets the colors and styles for type declarations and references."
    )),
    functions: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.functions",
      "Sets the colors and styles for functions declarations and references."
    )),
    variables: oe(r(
      "vs/workbench/services/themes/common/themeConfiguration",
      "editorColors.variables",
      "Sets the colors and styles for variables declarations and references."
    )),
    textMateRules: {
      description: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        "editorColors.textMateRules",
        "Sets colors and styles using textmate theming rules (advanced)."
      ),
      $ref: Ve
    },
    semanticHighlighting: {
      description: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        "editorColors.semanticHighlighting",
        "Whether semantic highlighting should be enabled for this theme."
      ),
      deprecationMessage: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        "editorColors.semanticHighlighting.deprecationMessage",
        "Use `enabled` in `editor.semanticTokenColorCustomizations` setting instead."
      ),
      markdownDeprecationMessage: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        { key: "editorColors.semanticHighlighting.deprecationMessageMarkdown", comment: ["{0} will become a link to another setting."] },
        "Use `enabled` in {0} setting instead.",
        W("editor.semanticTokenColorCustomizations")
      ),
      type: "boolean"
    }
  },
  additionalProperties: !1
}, lo = {
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "editorColors",
    "Overrides editor syntax colors and font style from the currently selected color theme."
  ),
  default: {},
  allOf: [{ ...ho, patternProperties: { "^\\[": {} } }]
}, st = {
  type: "object",
  properties: {
    enabled: {
      type: "boolean",
      description: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        "editorColors.semanticHighlighting.enabled",
        "Whether semantic highlighting is enabled or disabled for this theme"
      ),
      suggestSortText: "0_enabled"
    },
    rules: {
      $ref: Jt,
      description: r(
        "vs/workbench/services/themes/common/themeConfiguration",
        "editorColors.semanticHighlighting.rules",
        "Semantic token styling rules for this theme."
      ),
      suggestSortText: "0_rules"
    }
  },
  additionalProperties: !1
}, mo = {
  description: r(
    "vs/workbench/services/themes/common/themeConfiguration",
    "semanticTokenColors",
    "Overrides editor semantic token color and styles from the currently selected color theme."
  ),
  default: {},
  allOf: [{ ...st, patternProperties: { "^\\[": {} } }]
}, uo = {
  id: "editor",
  order: 7.2,
  type: "object",
  properties: {
    [S.TOKEN_COLOR_CUSTOMIZATIONS]: lo,
    [S.SEMANTIC_TOKEN_COLOR_CUSTOMIZATIONS]: mo
  }
};
ue.registerConfiguration(uo);
function Gt(s) {
  s.sort((c, a) => c.label.localeCompare(a.label)), ne.splice(0, ne.length, ...s.map((c) => c.settingsId)), se.splice(0, se.length, ...s.map((c) => c.description || "")), ie.splice(0, ie.length, ...s.map((c) => c.label || ""));
  const e = { properties: {} }, t = { properties: {} }, o = { properties: {} }, n = { $ref: ut, additionalProperties: !1 }, i = { properties: ho.properties, additionalProperties: !1 };
  for (const c of s) {
    const a = `[${c.settingsId}]`;
    e.properties[a] = n, t.properties[a] = i, o.properties[a] = st;
  }
  e.patternProperties = { [Qe]: n }, t.patternProperties = { [Qe]: i }, o.patternProperties = { [Qe]: st }, ao.allOf[1] = e, lo.allOf[1] = t, mo.allOf[1] = o, ue.notifyConfigurationSchemaUpdated(We, uo);
}
function Ut(s) {
  Le.enum.splice(1, Number.MAX_VALUE, ...s.map((e) => e.settingsId)), Le.enumItemLabels.splice(1, Number.MAX_VALUE, ...s.map((e) => e.label)), Le.enumDescriptions.splice(1, Number.MAX_VALUE, ...s.map((e) => e.description || "")), ue.notifyConfigurationSchemaUpdated(We);
}
function Bt(s) {
  He.enum.splice(1, Number.MAX_VALUE, ...s.map((e) => e.settingsId)), He.enumItemLabels.splice(1, Number.MAX_VALUE, ...s.map((e) => e.label)), He.enumDescriptions.splice(1, Number.MAX_VALUE, ...s.map((e) => e.description || "")), ue.notifyConfigurationSchemaUpdated(We);
}
class si {
  constructor(e) {
    this.configurationService = e;
  }
  get colorTheme() {
    return this.configurationService.getValue(S.COLOR_THEME);
  }
  get fileIconTheme() {
    return this.configurationService.getValue(S.FILE_ICON_THEME);
  }
  get productIconTheme() {
    return this.configurationService.getValue(S.PRODUCT_ICON_THEME);
  }
  get colorCustomizations() {
    return this.configurationService.getValue(S.COLOR_CUSTOMIZATIONS) || {};
  }
  get tokenColorCustomizations() {
    return this.configurationService.getValue(S.TOKEN_COLOR_CUSTOMIZATIONS) || {};
  }
  get semanticTokenColorCustomizations() {
    return this.configurationService.getValue(S.SEMANTIC_TOKEN_COLOR_CUSTOMIZATIONS);
  }
  async setColorTheme(e, t) {
    return await this.writeConfiguration(S.COLOR_THEME, e.settingsId, t), e;
  }
  async setFileIconTheme(e, t) {
    return await this.writeConfiguration(S.FILE_ICON_THEME, e.settingsId, t), e;
  }
  async setProductIconTheme(e, t) {
    return await this.writeConfiguration(S.PRODUCT_ICON_THEME, e.settingsId, t), e;
  }
  isDefaultColorTheme() {
    var t;
    const e = this.configurationService.inspect(S.COLOR_THEME);
    return e && ((t = e.default) == null ? void 0 : t.value) === e.value;
  }
  findAutoConfigurationTarget(e) {
    const t = this.configurationService.inspect(e);
    if (Z(t.workspaceFolderValue))
      if (Z(t.workspaceValue)) {
        if (!Z(t.userRemote))
          return 4;
      } else
        return 5;
    else
      return 6;
    return 2;
  }
  async writeConfiguration(e, t, o) {
    if (o === void 0 || o === "preview")
      return;
    const n = this.configurationService.inspect(e);
    if (o === "auto")
      return this.configurationService.updateValue(e, t);
    if (o === 2) {
      if (t === n.userValue)
        return Promise.resolve(void 0);
      if (t === n.defaultValue) {
        if (Z(n.userValue))
          return Promise.resolve(void 0);
        t = void 0;
      }
    } else if ((o === 5 || o === 6 || o === 4) && t === n.value)
      return Promise.resolve(void 0);
    return this.configurationService.updateValue(e, t, o);
  }
}
const Ie = "", L = class L {
  constructor(e, t, o) {
    this.iconThemeDocument = { iconDefinitions: /* @__PURE__ */ new Map() }, this.id = e, this.label = t, this.settingsId = o, this.isLoaded = !1;
  }
  getIcon(e) {
    return ai(e, this.iconThemeDocument);
  }
  ensureLoaded(e, t) {
    return this.isLoaded ? Promise.resolve(this.styleSheetContent) : this.load(e, t);
  }
  reload(e, t) {
    return this.load(e, t);
  }
  async load(e, t) {
    const o = this.location;
    if (!o)
      return Promise.resolve(this.styleSheetContent);
    const n = [];
    return this.iconThemeDocument = await ri(e, o, n), this.isLoaded = !0, n.length && t.error(r(
      "vs/workbench/services/themes/browser/productIconThemeData",
      "error.parseicondefs",
      `Problems processing product icons definitions in {0}:
{1}`,
      o.toString(),
      n.join(`
`)
    )), this.styleSheetContent;
  }
  static fromExtensionTheme(e, t, o) {
    const n = o.extensionId + "-" + e.id, i = e.label || at(e.path), c = e.id, a = new L(n, i, c);
    return a.description = e.description, a.location = t, a.extensionData = o, a.watch = e._watch, a.isLoaded = !1, a;
  }
  static createUnloadedTheme(e) {
    const t = new L(e, "", "__" + e);
    return t.isLoaded = !1, t.extensionData = void 0, t.watch = !1, t;
  }
  static get defaultTheme() {
    let e = L._defaultProductIconTheme;
    return e || (e = L._defaultProductIconTheme = new L(Ie, r(
      "vs/workbench/services/themes/browser/productIconThemeData",
      "defaultTheme",
      "Default"
    ), v.PRODUCT_ICON_THEME), e.isLoaded = !0, e.extensionData = void 0, e.watch = !1), e;
  }
  static fromStorageData(e) {
    const t = e.get(L.STORAGE_KEY, 0);
    if (t)
      try {
        const o = JSON.parse(t), n = new L("", "", "");
        for (const a in o)
          switch (a) {
            case "id":
            case "label":
            case "description":
            case "settingsId":
            case "styleSheetContent":
            case "watch":
              n[a] = o[a];
              break;
            case "location":
              break;
            case "extensionData":
              n.extensionData = j.fromJSONObject(o.extensionData);
              break;
          }
        const { iconDefinitions: i, iconFontDefinitions: c } = o;
        if (Array.isArray(i) && Me(c)) {
          const a = /* @__PURE__ */ new Map();
          for (const l of i) {
            const { id: d, fontCharacter: h, fontId: u } = l;
            if (M(d) && M(h))
              if (M(u)) {
                const f = xt.fromJSONObject(c[u]);
                f && a.set(d, { fontCharacter: h, font: { id: u, definition: f } });
              } else
                a.set(d, { fontCharacter: h });
          }
          n.iconThemeDocument = { iconDefinitions: a };
        }
        return n;
      } catch {
        return;
      }
  }
  toStorage(e) {
    const t = [], o = {};
    for (const i of this.iconThemeDocument.iconDefinitions.entries()) {
      const c = i[1].font;
      t.push({ id: i[0], fontCharacter: i[1].fontCharacter, fontId: c == null ? void 0 : c.id }), c && o[c.id] === void 0 && (o[c.id] = xt.toJSONObject(c.definition));
    }
    const n = JSON.stringify({
      id: this.id,
      label: this.label,
      description: this.description,
      settingsId: this.settingsId,
      styleSheetContent: this.styleSheetContent,
      watch: this.watch,
      extensionData: j.toJSONObject(this.extensionData),
      iconDefinitions: t,
      iconFontDefinitions: o
    });
    e.store(L.STORAGE_KEY, n, 0, 1);
  }
};
L.STORAGE_KEY = "productIconThemeData", L._defaultProductIconTheme = null;
let K = L;
function ri(s, e, t) {
  return s.readExtensionResource(e).then((o) => {
    const n = [], i = ht(o, n);
    if (n.length > 0)
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/browser/productIconThemeData",
        "error.cannotparseicontheme",
        "Problems parsing product icons file: {0}",
        n.map((h) => lt(h.error)).join(", ")
      )));
    if (mt(i) !== "object")
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/browser/productIconThemeData",
        "error.invalidformat",
        "Invalid format for product icons theme file: Object expected."
      )));
    if (!i.iconDefinitions || !Array.isArray(i.fonts) || !i.fonts.length)
      return Promise.reject(new Error(r(
        "vs/workbench/services/themes/browser/productIconThemeData",
        "error.missingProperties",
        "Invalid format for product icons theme file: Must contain iconDefinitions and fonts."
      )));
    const c = Fe(e), a = /* @__PURE__ */ new Map();
    for (const h of i.fonts)
      if (M(h.id) && h.id.match(St)) {
        const u = h.id;
        let f;
        M(h.weight) && h.weight.match(It) ? f = h.weight : t.push(r(
          "vs/workbench/services/themes/browser/productIconThemeData",
          "error.fontWeight",
          "Invalid font weight in font '{0}'. Ignoring setting.",
          h.id
        ));
        let m;
        M(h.style) && h.style.match(Ct) ? m = h.style : t.push(r(
          "vs/workbench/services/themes/browser/productIconThemeData",
          "error.fontStyle",
          "Invalid font style in font '{0}'. Ignoring setting.",
          h.id
        ));
        const p = [];
        if (Array.isArray(h.src))
          for (const b of h.src)
            if (M(b.path) && M(b.format) && b.format.match(Hn)) {
              const C = X(c, b.path);
              p.push({ location: C, format: b.format });
            } else
              t.push(r(
                "vs/workbench/services/themes/browser/productIconThemeData",
                "error.fontSrc",
                "Invalid font source in font '{0}'. Ignoring source.",
                h.id
              ));
        p.length ? a.set(u, { weight: f, style: m, src: p }) : t.push(r(
          "vs/workbench/services/themes/browser/productIconThemeData",
          "error.noFontSrc",
          "No valid font source in font '{0}'. Ignoring font definition.",
          h.id
        ));
      } else
        t.push(r(
          "vs/workbench/services/themes/browser/productIconThemeData",
          "error.fontId",
          "Missing or invalid font id '{0}'. Skipping font definition.",
          h.id
        ));
    const l = /* @__PURE__ */ new Map(), d = i.fonts[0].id;
    for (const h in i.iconDefinitions) {
      const u = i.iconDefinitions[h];
      if (M(u.fontCharacter)) {
        const f = u.fontId ?? d, m = a.get(f);
        if (m) {
          const p = { id: `pi-${f}`, definition: m };
          l.set(h, { fontCharacter: u.fontCharacter, font: p });
        } else
          t.push(r(
            "vs/workbench/services/themes/browser/productIconThemeData",
            "error.icon.font",
            "Skipping icon definition '{0}'. Unknown font.",
            h
          ));
      } else
        t.push(r(
          "vs/workbench/services/themes/browser/productIconThemeData",
          "error.icon.fontCharacter",
          "Skipping icon definition '{0}'. Unknown fontCharacter.",
          h
        ));
    }
    return { iconDefinitions: l };
  });
}
const ci = Uo();
function ai(s, e) {
  const t = e.iconDefinitions;
  let o = t.get(s.id), n = s.defaults;
  for (; !o && nt.isThemeIcon(n); ) {
    const i = ci.getIcon(n.id);
    if (i)
      o = t.get(i.id), n = i.defaults;
    else
      return;
  }
  if (o)
    return o;
  if (!nt.isThemeIcon(n))
    return n;
}
const Te = "osColorScheme", be = -1, Pe = "vscode-theme-defaults", Ze = "vscode.vscode-theme-seti-vs-seti", zt = "file-icons-enabled", hi = "contributedColorTheme", li = "contributedFileIconTheme", mi = "contributedProductIconTheme", qt = J.as(Bo.ThemingContribution);
function di(s) {
  switch (s) {
    case ze:
      return `vs ${Pe}-themes-light_vs-json`;
    case Ne:
      return `vs-dark ${Pe}-themes-dark_vs-json`;
    case Be:
      return `hc-black ${Pe}-themes-hc_black-json`;
    case Ue:
      return `hc-light ${Pe}-themes-hc_light-json`;
  }
  return s;
}
const ui = Vn(), fi = Wn(), gi = Jn();
let rt = class {
  constructor(e, t, o, n, i, c, a, l, d, h, u, f) {
    var Y;
    this.storageService = t, this.configurationService = o, this.telemetryService = n, this.environmentService = i, this.extensionResourceLoaderService = a, this.logService = d, this.hostColorService = h, this.userDataInitializationService = u, this.hasDefaultUpdated = !1, this.themeExtensionsActivated = /* @__PURE__ */ new Map(), this.container = l.mainContainer, this.settings = new si(o), this.colorThemeRegistry = new Ye(ui, z.fromExtensionTheme), this.colorThemeWatcher = new Xe(c, i, this.reloadCurrentColorTheme.bind(this)), this.onColorThemeChange = new Ce({ leakWarningThreshold: 400 }), this.currentColorTheme = z.createUnloadedTheme(""), this.colorThemeSequencer = new Je(), this.fileIconThemeWatcher = new Xe(
      c,
      i,
      this.reloadCurrentFileIconTheme.bind(this)
    ), this.fileIconThemeRegistry = new Ye(
      fi,
      q.fromExtensionTheme,
      !0,
      q.noIconTheme
    ), this.fileIconThemeLoader = new Un(a, f), this.onFileIconThemeChange = new Ce({ leakWarningThreshold: 400 }), this.currentFileIconTheme = q.createUnloadedTheme(""), this.fileIconThemeSequencer = new Je(), this.productIconThemeWatcher = new Xe(
      c,
      i,
      this.reloadCurrentProductIconTheme.bind(this)
    ), this.productIconThemeRegistry = new Ye(
      gi,
      K.fromExtensionTheme,
      !0,
      K.defaultTheme
    ), this.onProductIconThemeChange = new Ce(), this.currentProductIconTheme = K.createUnloadedTheme(""), this.productIconThemeSequencer = new Je();
    let m = z.fromStorageData(this.storageService);
    const p = this.settings.colorTheme;
    m && p !== m.settingsId && this.settings.isDefaultColorTheme() && (this.hasDefaultUpdated = m.settingsId === v.COLOR_THEME_DARK_OLD || m.settingsId === v.COLOR_THEME_LIGHT_OLD, m = void 0);
    const b = this.getPreferredColorScheme(), C = p === v.COLOR_THEME_LIGHT ? zo : p === v.COLOR_THEME_DARK ? qo : void 0;
    if (b && (m == null ? void 0 : m.type) !== b && this.storageService.get(Te, be) !== b && (m = z.createUnloadedThemeForThemeType(b, void 0)), !m) {
      const H = (Y = i.options) == null ? void 0 : Y.initialColorTheme;
      H && (m = z.createUnloadedThemeForThemeType(H.themeType, H.colors ?? C));
    }
    m || (m = z.createUnloadedThemeForThemeType(gt ? w.LIGHT : w.DARK, C)), m.setCustomizations(this.settings), this.applyTheme(m, void 0, !0);
    const T = q.fromStorageData(this.storageService);
    T && this.applyAndSetFileIconTheme(T, !0);
    const I = K.fromStorageData(this.storageService);
    I && this.applyAndSetProductIconTheme(I, !0), e.whenInstalledExtensionsRegistered().then((H) => {
      this.installConfigurationListener(), this.installPreferredSchemeListener(), this.installRegistryListeners(), this.initialize().catch($e);
    });
    const P = Qt();
    P.id = "codiconStyles";
    const y = Ko(this);
    function te() {
      P.textContent = y.getCSS();
    }
    const G = new Vo(te, 0);
    y.onDidChange(() => G.schedule()), G.schedule();
  }
  initialize() {
    const e = this.environmentService.extensionDevelopmentLocationURI, t = e && e.length === 1 ? e[0] : void 0, o = async () => {
      const c = this.colorThemeRegistry.findThemeByExtensionLocation(t);
      if (c.length) {
        const h = c.find((u) => u.type === this.currentColorTheme.type);
        return this.setColorTheme(h ? h.id : c[0].id, void 0);
      }
      const a = this.getPreferredColorScheme(), l = this.storageService.get(Te, be);
      if (a !== l && (this.storageService.store(Te, a, be, 0), a && this.currentColorTheme.type !== a))
        return this.applyPreferredColorTheme(a);
      let d = this.colorThemeRegistry.findThemeBySettingsId(this.settings.colorTheme, void 0);
      if (!d) {
        await this.userDataInitializationService.whenInitializationFinished();
        const h = this.currentColorTheme.type === w.LIGHT ? v.COLOR_THEME_LIGHT : v.COLOR_THEME_DARK;
        d = this.colorThemeRegistry.findThemeBySettingsId(this.settings.colorTheme, h);
      }
      return this.setColorTheme(d && d.id, void 0);
    }, n = async () => {
      const c = this.fileIconThemeRegistry.findThemeByExtensionLocation(t);
      if (c.length)
        return this.setFileIconTheme(c[0].id, 8);
      let a = this.fileIconThemeRegistry.findThemeBySettingsId(this.settings.fileIconTheme);
      return a || (await this.userDataInitializationService.whenInitializationFinished(), a = this.fileIconThemeRegistry.findThemeBySettingsId(this.settings.fileIconTheme)), this.setFileIconTheme(a ? a.id : Ze, void 0);
    }, i = async () => {
      const c = this.productIconThemeRegistry.findThemeByExtensionLocation(t);
      if (c.length)
        return this.setProductIconTheme(c[0].id, 8);
      let a = this.productIconThemeRegistry.findThemeBySettingsId(this.settings.productIconTheme);
      return a || (await this.userDataInitializationService.whenInitializationFinished(), a = this.productIconThemeRegistry.findThemeBySettingsId(this.settings.productIconTheme)), this.setProductIconTheme(a ? a.id : Ie, void 0);
    };
    return Promise.all([o(), n(), i()]);
  }
  installConfigurationListener() {
    this.configurationService.onDidChangeConfiguration((e) => {
      let t = null;
      const o = () => (t === null && (t = this.getPreferredColorScheme()), t);
      if (e.affectsConfiguration(S.COLOR_THEME) && this.restoreColorTheme(), (e.affectsConfiguration(S.DETECT_COLOR_SCHEME) || e.affectsConfiguration(S.DETECT_HC)) && this.handlePreferredSchemeUpdated(), e.affectsConfiguration(S.PREFERRED_DARK_THEME) && o() === w.DARK && this.applyPreferredColorTheme(w.DARK), e.affectsConfiguration(S.PREFERRED_LIGHT_THEME) && o() === w.LIGHT && this.applyPreferredColorTheme(w.LIGHT), e.affectsConfiguration(S.PREFERRED_HC_DARK_THEME) && o() === w.HIGH_CONTRAST_DARK && this.applyPreferredColorTheme(w.HIGH_CONTRAST_DARK), e.affectsConfiguration(S.PREFERRED_HC_LIGHT_THEME) && o() === w.HIGH_CONTRAST_LIGHT && this.applyPreferredColorTheme(w.HIGH_CONTRAST_LIGHT), e.affectsConfiguration(S.FILE_ICON_THEME) && this.restoreFileIconTheme(), e.affectsConfiguration(S.PRODUCT_ICON_THEME) && this.restoreProductIconTheme(), this.currentColorTheme) {
        let n = !1;
        e.affectsConfiguration(S.COLOR_CUSTOMIZATIONS) && (this.currentColorTheme.setCustomColors(this.settings.colorCustomizations), n = !0), e.affectsConfiguration(S.TOKEN_COLOR_CUSTOMIZATIONS) && (this.currentColorTheme.setCustomTokenColors(this.settings.tokenColorCustomizations), n = !0), e.affectsConfiguration(S.SEMANTIC_TOKEN_COLOR_CUSTOMIZATIONS) && (this.currentColorTheme.setCustomSemanticTokenColors(this.settings.semanticTokenColorCustomizations), n = !0), n && (this.updateDynamicCSSRules(this.currentColorTheme), this.onColorThemeChange.fire(this.currentColorTheme));
      }
    });
  }
  installRegistryListeners() {
    let e;
    this.colorThemeRegistry.onDidChange(async (n) => {
      if (Gt(n.themes), await this.restoreColorTheme())
        this.currentColorTheme.settingsId === v.COLOR_THEME_DARK && !Z(e) && await this.colorThemeRegistry.findThemeById(e) ? (await this.setColorTheme(e, "auto"), e = void 0) : n.added.some((i) => i.settingsId === this.currentColorTheme.settingsId) && await this.reloadCurrentColorTheme();
      else if (n.removed.some((i) => i.settingsId === this.currentColorTheme.settingsId)) {
        e = this.currentColorTheme.id;
        const i = this.colorThemeRegistry.findThemeBySettingsId(v.COLOR_THEME_DARK);
        await this.setColorTheme(i, "auto");
      }
    });
    let t;
    this.fileIconThemeRegistry.onDidChange(async (n) => {
      Ut(n.themes), await this.restoreFileIconTheme() ? this.currentFileIconTheme.id === Ze && !Z(t) && this.fileIconThemeRegistry.findThemeById(t) ? (await this.setFileIconTheme(t, "auto"), t = void 0) : n.added.some((i) => i.settingsId === this.currentFileIconTheme.settingsId) && await this.reloadCurrentFileIconTheme() : n.removed.some((i) => i.settingsId === this.currentFileIconTheme.settingsId) && (t = this.currentFileIconTheme.id, await this.setFileIconTheme(Ze, "auto"));
    });
    let o;
    return this.productIconThemeRegistry.onDidChange(async (n) => {
      Bt(n.themes), await this.restoreProductIconTheme() ? this.currentProductIconTheme.id === Ie && !Z(o) && this.productIconThemeRegistry.findThemeById(o) ? (await this.setProductIconTheme(o, "auto"), o = void 0) : n.added.some((i) => i.settingsId === this.currentProductIconTheme.settingsId) && await this.reloadCurrentProductIconTheme() : n.removed.some((i) => i.settingsId === this.currentProductIconTheme.settingsId) && (o = this.currentProductIconTheme.id, await this.setProductIconTheme(Ie, "auto"));
    }), Promise.all([this.getColorThemes(), this.getFileIconThemes(), this.getProductIconThemes()]).then(([n, i, c]) => {
      Gt(n), Ut(i), Bt(c);
    });
  }
  installPreferredSchemeListener() {
    this.hostColorService.onDidChangeColorScheme(() => this.handlePreferredSchemeUpdated());
  }
  async handlePreferredSchemeUpdated() {
    const e = this.getPreferredColorScheme(), t = this.storageService.get(Te, be);
    if (e !== t) {
      if (this.storageService.store(Te, e, be, 1), e)
        return t || (this.themeSettingIdBeforeSchemeSwitch = this.settings.colorTheme), this.applyPreferredColorTheme(e);
      if (t && this.themeSettingIdBeforeSchemeSwitch) {
        const o = this.colorThemeRegistry.findThemeBySettingsId(this.themeSettingIdBeforeSchemeSwitch, void 0);
        o && this.setColorTheme(o.id, "auto");
      }
    }
  }
  getPreferredColorScheme() {
    if (this.configurationService.getValue(S.DETECT_HC) && this.hostColorService.highContrast)
      return this.hostColorService.dark ? w.HIGH_CONTRAST_DARK : w.HIGH_CONTRAST_LIGHT;
    if (this.configurationService.getValue(S.DETECT_COLOR_SCHEME))
      return this.hostColorService.dark ? w.DARK : w.LIGHT;
  }
  async applyPreferredColorTheme(e) {
    let t;
    switch (e) {
      case w.LIGHT:
        t = S.PREFERRED_LIGHT_THEME;
        break;
      case w.HIGH_CONTRAST_DARK:
        t = S.PREFERRED_HC_DARK_THEME;
        break;
      case w.HIGH_CONTRAST_LIGHT:
        t = S.PREFERRED_HC_LIGHT_THEME;
        break;
      default:
        t = S.PREFERRED_DARK_THEME;
    }
    const o = this.configurationService.getValue(t);
    if (o && typeof o == "string") {
      const n = this.colorThemeRegistry.findThemeBySettingsId(o, void 0);
      if (n) {
        const i = this.settings.findAutoConfigurationTarget(t);
        return this.setColorTheme(n.id, i);
      }
    }
    return null;
  }
  hasUpdatedDefaultThemes() {
    return this.hasDefaultUpdated;
  }
  getColorTheme() {
    return this.currentColorTheme;
  }
  async getColorThemes() {
    return this.colorThemeRegistry.getThemes();
  }
  async getMarketplaceColorThemes(e, t, o) {
    const n = this.extensionResourceLoaderService.getExtensionGalleryResourceURL({ publisher: e, name: t, version: o }, "extension");
    if (n)
      try {
        const i = await this.extensionResourceLoaderService.readExtensionResource(X(n, "package.json"));
        return this.colorThemeRegistry.getMarketplaceThemes(JSON.parse(i), n, j.fromName(e, t));
      } catch (i) {
        this.logService.error("Problem loading themes from marketplace", i);
      }
    return [];
  }
  get onDidColorThemeChange() {
    return this.onColorThemeChange.event;
  }
  setColorTheme(e, t) {
    return this.colorThemeSequencer.queue(async () => this.internalSetColorTheme(e, t));
  }
  async internalSetColorTheme(e, t) {
    var i;
    if (!e)
      return null;
    const o = M(e) ? di(e) : e.id;
    if (this.currentColorTheme.isLoaded && o === this.currentColorTheme.id)
      return t !== "preview" && this.currentColorTheme.toStorage(this.storageService), this.settings.setColorTheme(this.currentColorTheme, t);
    let n = this.colorThemeRegistry.findThemeById(o);
    if (!n)
      if (e instanceof z)
        n = e;
      else
        return null;
    try {
      return await n.ensureLoaded(this.extensionResourceLoaderService), n.setCustomizations(this.settings), this.applyTheme(n, t);
    } catch (c) {
      throw new Error(r("vs/workbench/services/themes/browser/workbenchThemeService", "error.cannotloadtheme", "Unable to load {0}: {1}", (i = n.location) == null ? void 0 : i.toString(), c.message));
    }
  }
  reloadCurrentColorTheme() {
    return this.colorThemeSequencer.queue(async () => {
      var e;
      try {
        const t = this.colorThemeRegistry.findThemeBySettingsId(this.currentColorTheme.settingsId) || this.currentColorTheme;
        await t.reload(this.extensionResourceLoaderService), t.setCustomizations(this.settings), await this.applyTheme(t, void 0, !1);
      } catch {
        this.logService.info("Unable to reload {0}: {1}", (e = this.currentColorTheme.location) == null ? void 0 : e.toString());
      }
    });
  }
  async restoreColorTheme() {
    return this.colorThemeSequencer.queue(async () => {
      const e = this.settings.colorTheme, t = this.colorThemeRegistry.findThemeBySettingsId(e);
      return t ? (e !== this.currentColorTheme.settingsId ? await this.internalSetColorTheme(t.id, void 0) : t !== this.currentColorTheme && (await t.ensureLoaded(this.extensionResourceLoaderService), t.setCustomizations(this.settings), await this.applyTheme(t, void 0, !0)), !0) : !1;
    });
  }
  updateDynamicCSSRules(e) {
    const t = /* @__PURE__ */ new Set(), o = {
      addRule: (i) => {
        t.has(i) || t.add(i);
      }
    };
    o.addRule(".monaco-workbench { forced-color-adjust: none; }"), qt.getThemingParticipants().forEach((i) => i(e, o, this.environmentService));
    const n = [];
    for (const i of Wo().getColors()) {
      const c = e.getColor(i.id, !0);
      c && n.push(`${Jo(i.id)}: ${c.toString()};`);
    }
    o.addRule(`.monaco-workbench { ${n.join(`
`)} }`), et([...t].join(`
`), hi);
  }
  applyTheme(e, t, o = !1) {
    return this.updateDynamicCSSRules(e), this.currentColorTheme.id ? this.container.classList.remove(...this.currentColorTheme.classNames) : this.container.classList.remove(Ne, ze, Be, Ue), this.container.classList.add(...e.classNames), this.currentColorTheme.clearCaches(), this.currentColorTheme = e, this.colorThemingParticipantChangeListener || (this.colorThemingParticipantChangeListener = qt.onThemingParticipantAdded((n) => this.updateDynamicCSSRules(this.currentColorTheme))), this.colorThemeWatcher.update(e), this.sendTelemetry(e.id, e.extensionData, "color"), o ? Promise.resolve(null) : (this.onColorThemeChange.fire(this.currentColorTheme), e.isLoaded && t !== "preview" && e.toStorage(this.storageService), this.settings.setColorTheme(this.currentColorTheme, t));
  }
  sendTelemetry(e, t, o) {
    if (t) {
      const n = o + t.extensionId;
      this.themeExtensionsActivated.get(n) || (this.telemetryService.publicLog2("activatePlugin", {
        id: t.extensionId,
        name: t.extensionName,
        isBuiltin: t.extensionIsBuiltin,
        publisherDisplayName: t.extensionPublisher,
        themeId: e
      }), this.themeExtensionsActivated.set(n, !0));
    }
  }
  async getFileIconThemes() {
    return this.fileIconThemeRegistry.getThemes();
  }
  getFileIconTheme() {
    return this.currentFileIconTheme;
  }
  get onDidFileIconThemeChange() {
    return this.onFileIconThemeChange.event;
  }
  async setFileIconTheme(e, t) {
    return this.fileIconThemeSequencer.queue(async () => this.internalSetFileIconTheme(e, t));
  }
  async internalSetFileIconTheme(e, t) {
    e === void 0 && (e = "");
    const o = M(e) ? e : e.id;
    if (o !== this.currentFileIconTheme.id || !this.currentFileIconTheme.isLoaded) {
      let i = this.fileIconThemeRegistry.findThemeById(o);
      !i && e instanceof q && (i = e), i || (i = q.noIconTheme), await i.ensureLoaded(this.fileIconThemeLoader), this.applyAndSetFileIconTheme(i);
    }
    const n = this.currentFileIconTheme;
    return n.isLoaded && t !== "preview" && (!n.location || !Rt(n.location)) && n.toStorage(this.storageService), await this.settings.setFileIconTheme(this.currentFileIconTheme, t), n;
  }
  async getMarketplaceFileIconThemes(e, t, o) {
    const n = this.extensionResourceLoaderService.getExtensionGalleryResourceURL({ publisher: e, name: t, version: o }, "extension");
    if (n)
      try {
        const i = await this.extensionResourceLoaderService.readExtensionResource(X(n, "package.json"));
        return this.fileIconThemeRegistry.getMarketplaceThemes(JSON.parse(i), n, j.fromName(e, t));
      } catch (i) {
        this.logService.error("Problem loading themes from marketplace", i);
      }
    return [];
  }
  async reloadCurrentFileIconTheme() {
    return this.fileIconThemeSequencer.queue(async () => {
      await this.currentFileIconTheme.reload(this.fileIconThemeLoader), this.applyAndSetFileIconTheme(this.currentFileIconTheme);
    });
  }
  async restoreFileIconTheme() {
    return this.fileIconThemeSequencer.queue(async () => {
      const e = this.settings.fileIconTheme, t = this.fileIconThemeRegistry.findThemeBySettingsId(e);
      return t ? (e !== this.currentFileIconTheme.settingsId ? await this.internalSetFileIconTheme(t.id, void 0) : t !== this.currentFileIconTheme && (await t.ensureLoaded(this.fileIconThemeLoader), this.applyAndSetFileIconTheme(t, !0)), !0) : !1;
    });
  }
  applyAndSetFileIconTheme(e, t = !1) {
    this.currentFileIconTheme = e, et(e.styleSheetContent, li), e.id ? this.container.classList.add(zt) : this.container.classList.remove(zt), this.fileIconThemeWatcher.update(e), e.id && this.sendTelemetry(e.id, e.extensionData, "fileIcon"), t || this.onFileIconThemeChange.fire(this.currentFileIconTheme);
  }
  async getProductIconThemes() {
    return this.productIconThemeRegistry.getThemes();
  }
  getProductIconTheme() {
    return this.currentProductIconTheme;
  }
  get onDidProductIconThemeChange() {
    return this.onProductIconThemeChange.event;
  }
  async setProductIconTheme(e, t) {
    return this.productIconThemeSequencer.queue(async () => this.internalSetProductIconTheme(e, t));
  }
  async internalSetProductIconTheme(e, t) {
    e === void 0 && (e = "");
    const o = M(e) ? e : e.id;
    if (o !== this.currentProductIconTheme.id || !this.currentProductIconTheme.isLoaded) {
      let i = this.productIconThemeRegistry.findThemeById(o);
      !i && e instanceof K && (i = e), i || (i = K.defaultTheme), await i.ensureLoaded(this.extensionResourceLoaderService, this.logService), this.applyAndSetProductIconTheme(i);
    }
    const n = this.currentProductIconTheme;
    return n.isLoaded && t !== "preview" && (!n.location || !Rt(n.location)) && n.toStorage(this.storageService), await this.settings.setProductIconTheme(this.currentProductIconTheme, t), n;
  }
  async getMarketplaceProductIconThemes(e, t, o) {
    const n = this.extensionResourceLoaderService.getExtensionGalleryResourceURL({ publisher: e, name: t, version: o }, "extension");
    if (n)
      try {
        const i = await this.extensionResourceLoaderService.readExtensionResource(X(n, "package.json"));
        return this.productIconThemeRegistry.getMarketplaceThemes(JSON.parse(i), n, j.fromName(e, t));
      } catch (i) {
        this.logService.error("Problem loading themes from marketplace", i);
      }
    return [];
  }
  async reloadCurrentProductIconTheme() {
    return this.productIconThemeSequencer.queue(async () => {
      await this.currentProductIconTheme.reload(this.extensionResourceLoaderService, this.logService), this.applyAndSetProductIconTheme(this.currentProductIconTheme);
    });
  }
  async restoreProductIconTheme() {
    return this.productIconThemeSequencer.queue(async () => {
      const e = this.settings.productIconTheme, t = this.productIconThemeRegistry.findThemeBySettingsId(e);
      return t ? (e !== this.currentProductIconTheme.settingsId ? await this.internalSetProductIconTheme(t.id, void 0) : t !== this.currentProductIconTheme && (await t.ensureLoaded(this.extensionResourceLoaderService, this.logService), this.applyAndSetProductIconTheme(t, !0)), !0) : !1;
    });
  }
  applyAndSetProductIconTheme(e, t = !1) {
    this.currentProductIconTheme = e, et(e.styleSheetContent, mi), this.productIconThemeWatcher.update(e), e.id && this.sendTelemetry(e.id, e.extensionData, "productIcon"), t || this.onProductIconThemeChange.fire(this.currentProductIconTheme);
  }
};
rt = Ke([
  k(0, Yo),
  k(1, Zt),
  k(2, Xt),
  k(3, eo),
  k(4, Qo),
  k(5, Zo),
  k(6, pt),
  k(7, Xo),
  k(8, to),
  k(9, en),
  k(10, tn),
  k(11, on)
], rt);
class Xe {
  constructor(e, t, o) {
    this.fileService = e, this.environmentService = t, this.onUpdate = o;
  }
  update(e) {
    nn(e.location, this.watchedLocation) || (this.dispose(), e.location && (e.watch || this.environmentService.isExtensionDevelopment) && (this.watchedLocation = e.location, this.watcherDisposable = this.fileService.watch(e.location), this.fileService.onDidFilesChange((t) => {
      this.watchedLocation && t.contains(this.watchedLocation, 0) && this.onUpdate();
    })));
  }
  dispose() {
    this.watcherDisposable = _t(this.watcherDisposable), this.fileChangeListener = _t(this.fileChangeListener), this.watchedLocation = void 0;
  }
}
function et(s, e) {
  const t = qe.document.head.getElementsByClassName(e);
  if (t.length === 0) {
    const o = Qt();
    o.className = e, o.textContent = s;
  } else
    t[0].textContent = s;
}
Kn();
Gn();
Nn();
var Kt;
(function(s) {
  s[s.SAVE = 0] = "SAVE", s[s.DONT_SAVE = 1] = "DONT_SAVE", s[s.CANCEL = 2] = "CANCEL";
})(Kt || (Kt = {}));
var Se;
const pi = sn("theme-selection-manage-extension", rn.gear, r(
  "vs/workbench/contrib/themes/browser/themes.contribution",
  "manageExtensionIcon",
  "Icon for the 'Manage' action in the theme selection quick pick."
));
let je = class {
  constructor(e, t, o, n, i, c, a, l, d) {
    this.getMarketplaceColorThemes = e, this.marketplaceQuery = t, this.extensionGalleryService = o, this.extensionManagementService = n, this.quickInputService = i, this.logService = c, this.progressService = a, this.paneCompositeService = l, this.dialogService = d, this._marketplaceExtensions = /* @__PURE__ */ new Set(), this._marketplaceThemes = [], this._searchOngoing = !1, this._searchError = void 0, this._onDidChange = new Ce(), this._queryDelayer = new dn(200), this._installedExtensions = n.getInstalled().then((h) => {
      const u = /* @__PURE__ */ new Set();
      for (const f of h)
        u.add(f.identifier.id);
      return u;
    });
  }
  get themes() {
    return this._marketplaceThemes;
  }
  get onDidChange() {
    return this._onDidChange.event;
  }
  trigger(e) {
    this._tokenSource && (this._tokenSource.cancel(), this._tokenSource = void 0), this._queryDelayer.trigger(() => (this._tokenSource = new un(), this.doSearch(e, this._tokenSource.token)));
  }
  async doSearch(e, t) {
    this._searchOngoing = !0, this._onDidChange.fire();
    try {
      const o = await this._installedExtensions, n = { text: `${this.marketplaceQuery} ${e}`, pageSize: 20 }, i = await this.extensionGalleryService.query(n, t);
      for (let c = 0; c < i.total && c < 1 && !t.isCancellationRequested; c++) {
        const a = this._marketplaceThemes.length, l = c === 0 ? i.firstPage : await i.getPage(c, t), d = [], h = [];
        for (let f = 0; f < l.length && !t.isCancellationRequested; f++) {
          const m = l[f];
          !o.has(m.identifier.id) && !this._marketplaceExtensions.has(m.identifier.id) && (this._marketplaceExtensions.add(m.identifier.id), d.push(this.getMarketplaceColorThemes(m.publisher, m.name, m.version)), h.push(m));
        }
        const u = await Promise.all(d);
        for (let f = 0; f < u.length; f++) {
          const m = h[f];
          for (const p of u[f])
            this._marketplaceThemes.push({ id: p.id, theme: p, label: p.label, description: `${m.displayName}  ${m.publisherDisplayName}`, galleryExtension: m, buttons: [wt] });
        }
        a !== this._marketplaceThemes.length && (this._marketplaceThemes.sort((f, m) => f.label.localeCompare(m.label)), this._onDidChange.fire());
      }
    } catch (o) {
      fn(o) || (this.logService.error("Error while searching for themes:", o), this._searchError = "message" in o ? o.message : String(o));
    } finally {
      this._searchOngoing = !1, this._onDidChange.fire();
    }
  }
  openQuickPick(e, t, o) {
    let n;
    return new Promise((i, c) => {
      const a = this.quickInputService.createQuickPick();
      a.items = [], a.sortByLabel = !1, a.matchOnDescription = !0, a.buttons = [this.quickInputService.backButton], a.title = "Marketplace Themes", a.placeholder = r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "themes.selectMarketplaceTheme",
        "Type to Search More. Select to Install. Up/Down Keys to Preview"
      ), a.canSelectMany = !1, a.onDidChangeValue(() => this.trigger(a.value)), a.onDidAccept(async (l) => {
        const d = a.selectedItems[0];
        if (d != null && d.galleryExtension) {
          n = "selected", a.hide();
          const h = await this.installExtension(d.galleryExtension);
          o(h ? d.theme : t, !0);
        }
      }), a.onDidTriggerItemButton((l) => {
        var d, h;
        if (Ge(l.item)) {
          const u = (h = (d = l.item.theme) == null ? void 0 : d.extensionData) == null ? void 0 : h.extensionId;
          u ? ce(this.paneCompositeService, `@id:${u}`) : ce(this.paneCompositeService, `${this.marketplaceQuery} ${a.value}`);
        }
      }), a.onDidChangeActive((l) => {
        var d;
        n === void 0 && o((d = l[0]) == null ? void 0 : d.theme, !1);
      }), a.onDidHide(() => {
        n === void 0 && (o(t, !0), n = "cancelled"), a.dispose(), i(n);
      }), a.onDidTriggerButton((l) => {
        l === this.quickInputService.backButton && (n = "back", a.hide());
      }), this.onDidChange(() => {
        var u;
        let l = this.themes;
        this._searchOngoing ? l = l.concat({ label: "$(sync~spin) Searching for themes...", id: void 0, alwaysShow: !0 }) : l.length === 0 && this._searchError && (l = [{ label: `$(error) ${r(
          "vs/workbench/contrib/themes/browser/themes.contribution",
          "search.error",
          "Error while searching for themes: {0}",
          this._searchError
        )}`, id: void 0, alwaysShow: !0 }]);
        const d = (u = a.activeItems[0]) == null ? void 0 : u.id, h = d ? l.find((f) => Ge(f) && f.id === d) : void 0;
        a.items = l, h && (a.activeItems = [h]);
      }), this.trigger(e), a.show();
    });
  }
  async installExtension(e) {
    if (ce(this.paneCompositeService, `@id:${e.identifier.id}`), !(await this.dialogService.confirm({
      message: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "installExtension.confirm",
        "This will install extension '{0}' published by '{1}'. Do you want to continue?",
        e.displayName,
        e.publisherDisplayName
      ),
      primaryButton: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "installExtension.button.ok",
        "OK"
      )
    })).confirmed)
      return !1;
    try {
      return await this.progressService.withProgress({
        location: 15,
        title: r(
          "vs/workbench/contrib/themes/browser/themes.contribution",
          "installing extensions",
          "Installing Extension {0}...",
          e.displayName
        )
      }, async () => {
        await this.extensionManagementService.installFromGallery(e, {
          isMachineScoped: !1
        });
      }), !0;
    } catch (o) {
      return this.logService.error(`Problem installing extension ${e.identifier.id}`, o), !1;
    }
  }
  dispose() {
    this._tokenSource && (this._tokenSource.cancel(), this._tokenSource = void 0), this._queryDelayer.dispose(), this._marketplaceExtensions.clear(), this._marketplaceThemes.length = 0;
  }
};
je = Ke([
  k(2, Tt),
  k(3, pn),
  k(4, oo),
  k(5, to),
  k(6, Tn),
  k(7, no),
  k(8, bn)
], je);
let we = class {
  constructor(e, t, o, n, i, c, a, l, d, h, u) {
    this.installMessage = e, this.browseMessage = t, this.placeholderMessage = o, this.marketplaceTag = n, this.setTheme = i, this.getMarketplaceColorThemes = c, this.quickInputService = a, this.extensionGalleryService = l, this.paneCompositeService = d, this.extensionResourceLoaderService = h, this.instantiationService = u;
  }
  async openQuickPick(e, t) {
    let o;
    this.extensionGalleryService.isEnabled() && (this.extensionResourceLoaderService.supportsExtensionGalleryResources && this.browseMessage ? (o = this.instantiationService.createInstance(je, this.getMarketplaceColorThemes.bind(this), this.marketplaceTag), e = [...Vt(this.browseMessage), ...e]) : e = [...e, ...Vt(this.installMessage)]);
    let n;
    const i = (a, l) => {
      n && clearTimeout(n), n = qe.setTimeout(() => {
        n = void 0;
        const d = a ?? t;
        this.setTheme(d, l ? "auto" : "preview").then(void 0, (h) => {
          $e(h), this.setTheme(t, void 0);
        });
      }, l ? 0 : 200);
    }, c = (a) => new Promise((l, d) => {
      let h = !1;
      const u = e.findIndex((m) => Ge(m) && m.id === a), f = this.quickInputService.createQuickPick();
      f.items = e, f.placeholder = this.placeholderMessage, f.activeItems = [e[u]], f.canSelectMany = !1, f.matchOnDescription = !0, f.onDidAccept(async (m) => {
        h = !0;
        const p = f.selectedItems[0];
        !p || typeof p.id > "u" ? o ? await o.openQuickPick(f.value, t, i) === "back" && await c(void 0) : ce(this.paneCompositeService, `${this.marketplaceTag} ${f.value}`) : i(p.theme, !0), f.hide(), l();
      }), f.onDidChangeActive((m) => {
        var p;
        return i((p = m[0]) == null ? void 0 : p.theme, !1);
      }), f.onDidHide(() => {
        h || (i(t, !0), l()), f.dispose();
      }), f.onDidTriggerItemButton((m) => {
        var p, b;
        if (Ge(m.item)) {
          const C = (b = (p = m.item.theme) == null ? void 0 : p.extensionData) == null ? void 0 : b.extensionId;
          C ? ce(this.paneCompositeService, `@id:${C}`) : ce(this.paneCompositeService, `${this.marketplaceTag} ${f.value}`);
        }
      }), f.show();
    });
    await c(t.id), o == null || o.dispose();
  }
};
we = Ke([
  k(6, oo),
  k(7, Tt),
  k(8, no),
  k(9, pt),
  k(10, ke)
], we);
const yt = "workbench.action.selectTheme";
he(class extends le {
  constructor() {
    super({
      id: yt,
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "selectTheme.label",
        "Color Theme"
      ),
      category: de.Preferences,
      f1: !0,
      keybinding: {
        weight: 200,
        primary: cn(2089, 2098)
      }
    });
  }
  async run(s) {
    const e = s.get(ee), t = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "installColorThemes",
      "Install Additional Color Themes..."
    ), o = "$(plus) " + r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "browseColorThemes",
      "Browse Additional Color Themes..."
    ), n = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "themes.selectTheme",
      "Select Color Theme (Up/Down Keys to Preview)"
    ), i = "category:themes", c = (m, p) => e.setColorTheme(m, p), a = (m, p, b) => e.getMarketplaceColorThemes(m, p, b), d = s.get(ke).createInstance(we, t, o, n, i, c, a), h = await e.getColorThemes(), u = e.getColorTheme(), f = [
      ...ye(h.filter((m) => m.type === w.LIGHT), r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "themes.category.light",
        "light themes"
      )),
      ...ye(h.filter((m) => m.type === w.DARK), r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "themes.category.dark",
        "dark themes"
      )),
      ...ye(h.filter((m) => an(m.type)), r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "themes.category.hc",
        "high contrast themes"
      ))
    ];
    await d.openQuickPick(f, u);
  }
});
const fo = "workbench.action.selectIconTheme";
he(class extends le {
  constructor() {
    super({
      id: fo,
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "selectIconTheme.label",
        "File Icon Theme"
      ),
      category: de.Preferences,
      f1: !0
    });
  }
  async run(s) {
    const e = s.get(ee), t = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "installIconThemes",
      "Install Additional File Icon Themes..."
    ), o = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "themes.selectIconTheme",
      "Select File Icon Theme (Up/Down Keys to Preview)"
    ), n = "tag:icon-theme", i = (h, u) => e.setFileIconTheme(h, u), c = (h, u, f) => e.getMarketplaceFileIconThemes(h, u, f), l = s.get(ke).createInstance(we, t, void 0, o, n, i, c), d = [
      { type: "separator", label: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "fileIconThemeCategory",
        "file icon themes"
      ) },
      { id: "", theme: q.noIconTheme, label: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "noIconThemeLabel",
        "None"
      ), description: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "noIconThemeDesc",
        "Disable File Icons"
      ) },
      ...ye(await e.getFileIconThemes())
    ];
    await l.openQuickPick(d, e.getFileIconTheme());
  }
});
const go = "workbench.action.selectProductIconTheme";
he(class extends le {
  constructor() {
    super({
      id: go,
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "selectProductIconTheme.label",
        "Product Icon Theme"
      ),
      category: de.Preferences,
      f1: !0
    });
  }
  async run(s) {
    const e = s.get(ee), t = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "installProductIconThemes",
      "Install Additional Product Icon Themes..."
    ), o = "$(plus) " + r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "browseProductIconThemes",
      "Browse Additional Product Icon Themes..."
    ), n = r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "themes.selectProductIconTheme",
      "Select Product Icon Theme (Up/Down Keys to Preview)"
    ), i = "tag:product-icon-theme", c = (u, f) => e.setProductIconTheme(u, f), a = (u, f, m) => e.getMarketplaceProductIconThemes(u, f, m), d = s.get(ke).createInstance(we, t, o, n, i, c, a), h = [
      { type: "separator", label: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "productIconThemeCategory",
        "product icon themes"
      ) },
      { id: Ie, theme: K.defaultTheme, label: r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "defaultProductIconThemeLabel",
        "Default"
      ) },
      ...ye(await e.getProductIconThemes())
    ];
    await d.openQuickPick(h, e.getProductIconTheme());
  }
});
hn.registerCommand("workbench.action.previewColorTheme", async function(s, e, t) {
  const o = s.get(ee);
  let n = Ti(await o.getColorThemes(), e);
  n.length === 0 && (n = await o.getMarketplaceColorThemes(e.publisher, e.name, e.version));
  for (const i of n)
    if (!t || i.settingsId === t)
      return await o.setColorTheme(i, "preview"), i.settingsId;
});
function Ti(s, e) {
  return s.filter(({ extensionData: t }) => t && t.extensionIsBuiltin && Dt(t.extensionPublisher, e.publisher) && Dt(t.extensionName, e.name));
}
function Vt(s) {
  return [
    {
      type: "separator"
    },
    {
      id: void 0,
      label: s,
      alwaysShow: !0,
      buttons: [wt]
    }
  ];
}
function ce(s, e) {
  return s.openPaneComposite(gn, 0, !0).then((t) => {
    t && ((t == null ? void 0 : t.getViewPaneContainer()).search(e), t.focus());
  });
}
function Ge(s) {
  return s.type !== "separator";
}
function bi(s) {
  const e = s.settingsId ?? void 0, t = {
    id: s.id,
    theme: s,
    label: s.label,
    description: s.description || (s.label === e ? void 0 : e)
  };
  return s.extensionData && (t.buttons = [wt]), t;
}
function ye(s, e) {
  const t = (n, i) => n.label.localeCompare(i.label), o = s.map(bi).sort(t);
  return o.length > 0 && e && o.unshift({ type: "separator", label: e }), o;
}
const wt = {
  iconClass: nt.asClassName(pi),
  tooltip: r(
    "vs/workbench/contrib/themes/browser/themes.contribution",
    "manage extension",
    "Manage Extension"
  )
};
he(class extends le {
  constructor() {
    super({
      id: "workbench.action.generateColorTheme",
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "generateColorTheme.label",
        "Generate Color Theme From Current Settings"
      ),
      category: de.Developer,
      f1: !0
    });
  }
  run(s) {
    const t = s.get(ee).getColorTheme(), n = J.as(Wt.ColorContribution).getColors().map((h) => h.id).sort(), i = {}, c = [];
    for (const h of n) {
      const u = t.getColor(h, !1);
      u ? i[h] = V.Format.CSS.formatHexA(u, !0) : c.push(h);
    }
    const a = [];
    for (const h of c) {
      const u = t.getColor(h);
      u ? i["__" + h] = V.Format.CSS.formatHexA(u, !0) : a.push(h);
    }
    for (const h of a)
      i["__" + h] = null;
    let l = JSON.stringify({
      $schema: co,
      type: t.type,
      colors: i,
      tokenColors: t.tokenColors.filter((h) => !!h.scope)
    }, null, "	");
    return l = l.replace(/\"__/g, '//"'), s.get(ln).openEditor({ resource: void 0, contents: l, languageId: "jsonc", options: { pinned: !0 } });
  }
});
const Si = "workbench.action.toggleLightDarkThemes";
he(class extends le {
  constructor() {
    super({
      id: Si,
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "toggleLightDarkThemes.label",
        "Toggle between Light/Dark Themes"
      ),
      category: de.Preferences,
      f1: !0
    });
  }
  async run(s) {
    const e = s.get(ee), t = s.get(Xt), o = e.getColorTheme();
    let n = S.PREFERRED_DARK_THEME;
    switch (o.type) {
      case w.LIGHT:
        n = S.PREFERRED_DARK_THEME;
        break;
      case w.DARK:
        n = S.PREFERRED_LIGHT_THEME;
        break;
      case w.HIGH_CONTRAST_LIGHT:
        n = S.PREFERRED_HC_DARK_THEME;
        break;
      case w.HIGH_CONTRAST_DARK:
        n = S.PREFERRED_HC_LIGHT_THEME;
        break;
    }
    const i = t.getValue(n);
    if (i && typeof i == "string") {
      const c = (await e.getColorThemes()).find((a) => a.settingsId === i);
      c && e.setColorTheme(c.id, "auto");
    }
  }
});
const Ci = "workbench.action.browseColorThemesInMarketplace";
he(class extends le {
  constructor() {
    super({
      id: Ci,
      title: me(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        "browseColorThemeInMarketPlace.label",
        "Browse Color Themes in Marketplace"
      ),
      category: de.Preferences,
      f1: !0
    });
  }
  async run(s) {
    const e = "category:themes", t = s.get(ee), o = s.get(Tt), n = s.get(pt), i = s.get(ke);
    if (!o.isEnabled() || !n.supportsExtensionGalleryResources)
      return;
    const c = t.getColorTheme(), a = (u, f, m) => t.getMarketplaceColorThemes(u, f, m);
    let l;
    const d = (u, f) => {
      l && clearTimeout(l), l = qe.setTimeout(() => {
        l = void 0;
        const m = u ?? c;
        t.setColorTheme(m, f ? "auto" : "preview").then(void 0, (p) => {
          $e(p), t.setColorTheme(c, void 0);
        });
      }, f ? 0 : 200);
    };
    await i.createInstance(je, a, e).openQuickPick("", t.getColorTheme(), d).then(void 0, $e);
  }
});
const Ee = new bt("ThemesSubMenu");
ve.appendMenuItem(bt.GlobalActivity, {
  title: r(
    "vs/workbench/contrib/themes/browser/themes.contribution",
    "themes",
    "Themes"
  ),
  submenu: Ee,
  group: "2_configuration",
  order: 7
});
ve.appendMenuItem(bt.MenubarPreferencesMenu, {
  title: r(
    "vs/workbench/contrib/themes/browser/themes.contribution",
    { key: "miSelectTheme", comment: ["&& denotes a mnemonic"] },
    "&&Theme"
  ),
  submenu: Ee,
  group: "2_configuration",
  order: 7
});
ve.appendMenuItem(Ee, {
  command: {
    id: yt,
    title: r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "selectTheme.label",
      "Color Theme"
    )
  },
  order: 1
});
ve.appendMenuItem(Ee, {
  command: {
    id: fo,
    title: r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "themes.selectIconTheme.label",
      "File Icon Theme"
    )
  },
  order: 2
});
ve.appendMenuItem(Ee, {
  command: {
    id: go,
    title: r(
      "vs/workbench/contrib/themes/browser/themes.contribution",
      "themes.selectProductIconTheme.label",
      "Product Icon Theme"
    )
  },
  order: 3
});
var ae;
let ct = (ae = class {
  constructor(e, t, o, n, i, c) {
    this._notificationService = e, this._workbenchThemeService = t, this._storageService = o, this._commandService = n, this._telemetryService = i, this._hostService = c, !o.getBoolean(Se.STORAGE_KEY, -1) && setTimeout(async () => {
      if (!o.getBoolean(Se.STORAGE_KEY, -1) && await this._hostService.hadLastFocus())
        if (this._storageService.store(Se.STORAGE_KEY, !0, -1, 0), this._workbenchThemeService.hasUpdatedDefaultThemes())
          this._showYouGotMigratedNotification();
        else {
          const a = this._workbenchThemeService.getColorTheme().settingsId;
          (a === v.COLOR_THEME_LIGHT_OLD || a === v.COLOR_THEME_DARK_OLD) && this._tryNewThemeNotification();
        }
    }, 3e3);
  }
  async _showYouGotMigratedNotification() {
    const e = this._workbenchThemeService.getColorTheme().type === w.LIGHT, t = e ? v.COLOR_THEME_LIGHT : v.COLOR_THEME_DARK, o = (await this._workbenchThemeService.getColorThemes()).find((n) => n.settingsId === t);
    if (o) {
      const n = [
        {
          label: r(
            "vs/workbench/contrib/themes/browser/themes.contribution",
            "button.keep",
            "Keep New Theme"
          ),
          run: () => {
            this._writeTelemetry("keepNew");
          }
        },
        {
          label: r(
            "vs/workbench/contrib/themes/browser/themes.contribution",
            "button.browse",
            "Browse Themes"
          ),
          run: () => {
            this._writeTelemetry("browse"), this._commandService.executeCommand(yt);
          }
        },
        {
          label: r(
            "vs/workbench/contrib/themes/browser/themes.contribution",
            "button.revert",
            "Revert"
          ),
          run: async () => {
            this._writeTelemetry("keepOld");
            const i = e ? v.COLOR_THEME_LIGHT_OLD : v.COLOR_THEME_DARK_OLD, c = (await this._workbenchThemeService.getColorThemes()).find((a) => a.settingsId === i);
            c && this._workbenchThemeService.setColorTheme(c, "auto");
          }
        }
      ];
      await this._notificationService.prompt(Pt.Info, r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        { key: "themeUpdatedNotification", comment: ["{0} is the name of the new default theme"] },
        "Visual Studio Code now ships with a new default theme '{0}'. If you prefer, you can switch back to the old theme or try one of the many other color themes available.",
        o.label
      ), n, {
        onCancel: () => this._writeTelemetry("cancel")
      });
    }
  }
  async _tryNewThemeNotification() {
    const e = this._workbenchThemeService.getColorTheme().type === w.LIGHT ? v.COLOR_THEME_LIGHT : v.COLOR_THEME_DARK, t = (await this._workbenchThemeService.getColorThemes()).find((o) => o.settingsId === e);
    if (t) {
      const o = [
        {
          label: r(
            "vs/workbench/contrib/themes/browser/themes.contribution",
            "button.tryTheme",
            "Try New Theme"
          ),
          run: () => {
            this._writeTelemetry("tryNew"), this._workbenchThemeService.setColorTheme(t, "auto");
          }
        },
        {
          label: r(
            "vs/workbench/contrib/themes/browser/themes.contribution",
            "button.cancel",
            "Cancel"
          ),
          run: () => {
            this._writeTelemetry("cancel");
          }
        }
      ];
      await this._notificationService.prompt(Pt.Info, r(
        "vs/workbench/contrib/themes/browser/themes.contribution",
        { key: "newThemeNotification", comment: ["{0} is the name of the new default theme"] },
        "Visual Studio Code now ships with a new default theme '{0}'. Do you want to give it a try?",
        t.label
      ), o, { onCancel: () => this._writeTelemetry("cancel") });
    }
  }
  _writeTelemetry(e) {
    this._telemetryService.publicLog2("themeUpdatedNotication", {
      web: gt,
      reaction: e
    });
  }
}, Se = ae, ae.STORAGE_KEY = "themeUpdatedNotificationShown", ae);
ct = Se = Ke([
  k(0, Sn),
  k(1, ee),
  k(2, Zt),
  k(3, Cn),
  k(4, eo),
  k(5, In)
], ct);
const Ii = J.as(mn.Workbench);
Ii.registerWorkbenchContribution(ct, 4);
class yi extends rt {
  registerEditorContainer() {
    return {
      dispose() {
      }
    };
  }
  setTheme(e) {
    this.getColorThemes().then((t) => {
      setTimeout(() => {
        this.setColorTheme(t.find((o) => o.settingsId === e) ?? e, 8);
      });
    });
  }
}
function xi() {
  return {
    ...yn(),
    [wn.toString()]: new Yt(yi, [], !1)
  };
}
export {
  xi as default
};
