// vscode/vscode/src/vs/workbench/contrib/files/common/dirtyFilesIndicator.js
import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { localizeWithPath } from '../../../../nls.js';
import { VIEWLET_ID } from './files.js';
import { ILifecycleService } from '../../../services/lifecycle/common/lifecycle.js';
import { Disposable, MutableDisposable } from '../../../../base/common/lifecycle.js';
import { NumberBadge, IActivityService } from '../../../services/activity/common/activity.js';
import { IWorkingCopyService } from '../../../services/workingCopy/common/workingCopyService.js';
import { IFilesConfigurationService } from '../../../services/filesConfiguration/common/filesConfigurationService.js';
let DirtyFilesIndicator = class DirtyFilesIndicator extends Disposable {
    static { this.ID = 'workbench.contrib.dirtyFilesIndicator'; }
    constructor(lifecycleService, activityService, workingCopyService, filesConfigurationService) {
        super();
        this.lifecycleService = lifecycleService;
        this.activityService = activityService;
        this.workingCopyService = workingCopyService;
        this.filesConfigurationService = filesConfigurationService;
        this.badgeHandle = this._register(( new MutableDisposable()));
        this.lastKnownDirtyCount = 0;
        this.updateActivityBadge();
        this.registerListeners();
    }
    registerListeners() {
        this._register(this.workingCopyService.onDidChangeDirty(workingCopy => this.onWorkingCopyDidChangeDirty(workingCopy)));
        this.lifecycleService.onDidShutdown(() => this.dispose());
    }
    onWorkingCopyDidChangeDirty(workingCopy) {
        const gotDirty = workingCopy.isDirty();
        if (gotDirty && !((workingCopy.capabilities & 2) ) && this.filesConfigurationService.hasShortAutoSaveDelay(workingCopy.resource)) {
            return;
        }
        if (gotDirty || this.lastKnownDirtyCount > 0) {
            this.updateActivityBadge();
        }
    }
    updateActivityBadge() {
        const dirtyCount = this.lastKnownDirtyCount = this.workingCopyService.dirtyCount;
        if (dirtyCount > 0) {
            this.badgeHandle.value = this.activityService.showViewContainerActivity(VIEWLET_ID, {
                badge: ( new NumberBadge(dirtyCount, num => num === 1 ? ( localizeWithPath(
                    'vs/workbench/contrib/files/common/dirtyFilesIndicator',
                    'dirtyFile',
                    "1 unsaved file"
                )) : ( localizeWithPath(
                    'vs/workbench/contrib/files/common/dirtyFilesIndicator',
                    'dirtyFiles',
                    "{0} unsaved files",
                    dirtyCount
                )))),
            });
        }
        else {
            this.badgeHandle.clear();
        }
    }
};
DirtyFilesIndicator = ( __decorate([
    ( __param(0, ILifecycleService)),
    ( __param(1, IActivityService)),
    ( __param(2, IWorkingCopyService)),
    ( __param(3, IFilesConfigurationService))
], DirtyFilesIndicator));
export { DirtyFilesIndicator };
