// monaco-languageclient/lib/vscode/services.js
/* --------------------------------------------------------------------------------------------
 * Copyright (c) 2024 TypeFox and others.
 * Licensed under the MIT License. See LICENSE in the package root for license information.
 * ------------------------------------------------------------------------------------------ */
import 'vscode/localExtensionHost';
import { ILogService, initialize, StandaloneServices } from 'vscode/services';
import getLanguagesServiceOverride from '@codingame/monaco-vscode-languages-service-override';
import getModelServiceOverride from '@codingame/monaco-vscode-model-service-override';
import { FakeWorker as Worker } from './fakeWorker.js';
export const initEnhancedMonacoEnvironment = () => {
    const monWin = self;
    if (!monWin.MonacoEnvironment) {
        monWin.MonacoEnvironment = {};
    }
    const envEnhanced = monWin.MonacoEnvironment;
    if (envEnhanced.vscodeApiInitialised === undefined) {
        envEnhanced.vscodeApiInitialised = false;
    }
    if (envEnhanced.vscodeInitialising === undefined) {
        envEnhanced.vscodeInitialising = false;
    }
    return envEnhanced;
};
export const supplyRequiredServices = async () => {
    return {
        ...getLanguagesServiceOverride(),
        ...getModelServiceOverride()
    };
};
export const reportServiceLoading = (services, debugLogging) => {
    for (const serviceName of Object.keys(services)) {
        if (debugLogging) {
            console.log(`Loading service: ${serviceName}`);
        }
    }
};
export const mergeServices = (services, overrideServices) => {
    for (const [name, service] of Object.entries(services)) {
        overrideServices[name] = service;
    }
};
export const initServices = async (config, caller, performChecks) => {
    const envEnhanced = initEnhancedMonacoEnvironment();
    if (!envEnhanced.vscodeInitialising) {
        if (envEnhanced.vscodeApiInitialised) {
            if (config?.debugLogging === true) {
                console.log('Initialization of vscode services can only performed once!');
            }
        }
        else {
            envEnhanced.vscodeInitialising = true;
            if (config?.debugLogging === true) {
                console.log(`Initializing vscode services. Caller: ${caller ?? 'unknown'}`);
            }
            await importAllServices(config, performChecks);
            if (config?.debugLogging === true) {
                console.log('Initialization of vscode services completed successfully.');
            }
            envEnhanced.vscodeApiInitialised = true;
        }
    }
};
/**
 * monaco-vscode-api automatically loads the following services:
 *  - layout
 *  - environment
 *  - extension
 *  - files
 *  - quickAccess
 * monaco-languageclient always adds the following services:
 *   - languages
 *   - model
 */
export const importAllServices = async (config, performChecks) => {
    const lc = config ?? {};
    const userServices = lc.userServices ?? {};
    const lcRequiredServices = await supplyRequiredServices();
    mergeServices(lcRequiredServices, userServices);
    await configureExtHostWorker(config?.enableExtHostWorker === true, userServices);
    reportServiceLoading(userServices, lc.debugLogging === true);
    if (performChecks === undefined || performChecks()) {
        await initialize(userServices);
        const logLevel = lc.workspaceConfig?.developmentOptions?.logLevel;
        if (logLevel) {
            StandaloneServices.get(ILogService).setLevel(logLevel);
        }
    }
};
/**
 * Enable ext host to run in a worker
 */
export const configureExtHostWorker = async (enableExtHostWorker, userServices) => {
    if (enableExtHostWorker) {
        const fakeWorker = new Worker(new URL('vscode/workers/extensionHost.worker', import.meta.url), { type: 'module' });
        const workerConfig = {
            url: fakeWorker.url.toString(),
            options: fakeWorker.options
        };
        const getExtensionServiceOverride = (await import('@codingame/monaco-vscode-extensions-service-override')).default;
        const extHostServices = {
            ...getExtensionServiceOverride(workerConfig),
        };
        mergeServices(extHostServices, userServices);
    }
};
