// vscode/vscode/src/vs/workbench/contrib/preferences/browser/settingsLayout.js
import { localizeWithPath } from '../../../../nls.js';
const defaultCommonlyUsedSettings = [
    'files.autoSave',
    'editor.fontSize',
    'editor.fontFamily',
    'editor.tabSize',
    'editor.renderWhitespace',
    'editor.cursorStyle',
    'editor.multiCursorModifier',
    'editor.insertSpaces',
    'editor.wordWrap',
    'files.exclude',
    'files.associations',
    'workbench.editor.enablePreview'
];
function getCommonlyUsedData(toggleData) {
    return {
        id: 'commonlyUsed',
        label: ( localizeWithPath(
            'vs/workbench/contrib/preferences/browser/settingsLayout',
            'commonlyUsed',
            "Commonly Used"
        )),
        settings: toggleData?.commonlyUsed ?? defaultCommonlyUsedSettings
    };
}
const tocData = {
    id: 'root',
    label: 'root',
    children: [
        {
            id: 'editor',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'textEditor',
                "Text Editor"
            )),
            settings: ['editor.*'],
            children: [
                {
                    id: 'editor/cursor',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'cursor',
                        "Cursor"
                    )),
                    settings: ['editor.cursor*']
                },
                {
                    id: 'editor/find',
                    label: ( localizeWithPath('vs/workbench/contrib/preferences/browser/settingsLayout', 'find', "Find")),
                    settings: ['editor.find.*']
                },
                {
                    id: 'editor/font',
                    label: ( localizeWithPath('vs/workbench/contrib/preferences/browser/settingsLayout', 'font', "Font")),
                    settings: ['editor.font*']
                },
                {
                    id: 'editor/format',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'formatting',
                        "Formatting"
                    )),
                    settings: ['editor.format*']
                },
                {
                    id: 'editor/diffEditor',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'diffEditor',
                        "Diff Editor"
                    )),
                    settings: ['diffEditor.*']
                },
                {
                    id: 'editor/multiDiffEditor',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'multiDiffEditor',
                        "Multi-File Diff Editor"
                    )),
                    settings: ['multiDiffEditor.*']
                },
                {
                    id: 'editor/minimap',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'minimap',
                        "Minimap"
                    )),
                    settings: ['editor.minimap.*']
                },
                {
                    id: 'editor/suggestions',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'suggestions',
                        "Suggestions"
                    )),
                    settings: ['editor.*suggest*']
                },
                {
                    id: 'editor/files',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'files',
                        "Files"
                    )),
                    settings: ['files.*']
                }
            ]
        },
        {
            id: 'workbench',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'workbench',
                "Workbench"
            )),
            settings: ['workbench.*'],
            children: [
                {
                    id: 'workbench/appearance',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'appearance',
                        "Appearance"
                    )),
                    settings: ['workbench.activityBar.*', 'workbench.*color*', 'workbench.fontAliasing', 'workbench.iconTheme', 'workbench.sidebar.location', 'workbench.*.visible', 'workbench.tips.enabled', 'workbench.tree.*', 'workbench.view.*']
                },
                {
                    id: 'workbench/breadcrumbs',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'breadcrumbs',
                        "Breadcrumbs"
                    )),
                    settings: ['breadcrumbs.*']
                },
                {
                    id: 'workbench/editor',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'editorManagement',
                        "Editor Management"
                    )),
                    settings: ['workbench.editor.*']
                },
                {
                    id: 'workbench/settings',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'settings',
                        "Settings Editor"
                    )),
                    settings: ['workbench.settings.*']
                },
                {
                    id: 'workbench/zenmode',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'zenMode',
                        "Zen Mode"
                    )),
                    settings: ['zenmode.*']
                },
                {
                    id: 'workbench/screencastmode',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'screencastMode',
                        "Screencast Mode"
                    )),
                    settings: ['screencastMode.*']
                }
            ]
        },
        {
            id: 'window',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'window',
                "Window"
            )),
            settings: ['window.*'],
            children: [
                {
                    id: 'window/newWindow',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'newWindow',
                        "New Window"
                    )),
                    settings: ['window.*newwindow*']
                }
            ]
        },
        {
            id: 'features',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'features',
                "Features"
            )),
            children: [
                {
                    id: 'features/accessibilitySignals',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'accessibility.signals',
                        'Accessibility Signals'
                    )),
                    settings: ['accessibility.signals.*', 'audioCues.*']
                },
                {
                    id: 'features/accessibility',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'accessibility',
                        "Accessibility"
                    )),
                    settings: ['accessibility.*']
                },
                {
                    id: 'features/explorer',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'fileExplorer',
                        "Explorer"
                    )),
                    settings: ['explorer.*', 'outline.*']
                },
                {
                    id: 'features/search',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'search',
                        "Search"
                    )),
                    settings: ['search.*']
                },
                {
                    id: 'features/debug',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'debug',
                        "Debug"
                    )),
                    settings: ['debug.*', 'launch']
                },
                {
                    id: 'features/testing',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'testing',
                        "Testing"
                    )),
                    settings: ['testing.*']
                },
                {
                    id: 'features/scm',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'scm',
                        "Source Control"
                    )),
                    settings: ['scm.*']
                },
                {
                    id: 'features/extensions',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'extensions',
                        "Extensions"
                    )),
                    settings: ['extensions.*']
                },
                {
                    id: 'features/terminal',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'terminal',
                        "Terminal"
                    )),
                    settings: ['terminal.*']
                },
                {
                    id: 'features/task',
                    label: ( localizeWithPath('vs/workbench/contrib/preferences/browser/settingsLayout', 'task', "Task")),
                    settings: ['task.*']
                },
                {
                    id: 'features/problems',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'problems',
                        "Problems"
                    )),
                    settings: ['problems.*']
                },
                {
                    id: 'features/output',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'output',
                        "Output"
                    )),
                    settings: ['output.*']
                },
                {
                    id: 'features/comments',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'comments',
                        "Comments"
                    )),
                    settings: ['comments.*']
                },
                {
                    id: 'features/remote',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'remote',
                        "Remote"
                    )),
                    settings: ['remote.*']
                },
                {
                    id: 'features/timeline',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'timeline',
                        "Timeline"
                    )),
                    settings: ['timeline.*']
                },
                {
                    id: 'features/notebook',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'notebook',
                        'Notebook'
                    )),
                    settings: ['notebook.*', 'interactiveWindow.*']
                },
                {
                    id: 'features/mergeEditor',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'mergeEditor',
                        'Merge Editor'
                    )),
                    settings: ['mergeEditor.*']
                },
                {
                    id: 'features/chat',
                    label: ( localizeWithPath('vs/workbench/contrib/preferences/browser/settingsLayout', 'chat', 'Chat')),
                    settings: ['chat.*', 'inlineChat.*']
                }
            ]
        },
        {
            id: 'application',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'application',
                "Application"
            )),
            children: [
                {
                    id: 'application/http',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'proxy',
                        "Proxy"
                    )),
                    settings: ['http.*']
                },
                {
                    id: 'application/keyboard',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'keyboard',
                        "Keyboard"
                    )),
                    settings: ['keyboard.*']
                },
                {
                    id: 'application/update',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'update',
                        "Update"
                    )),
                    settings: ['update.*']
                },
                {
                    id: 'application/telemetry',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'telemetry',
                        "Telemetry"
                    )),
                    settings: ['telemetry.*']
                },
                {
                    id: 'application/settingsSync',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'settingsSync',
                        "Settings Sync"
                    )),
                    settings: ['settingsSync.*']
                },
                {
                    id: 'application/experimental',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'experimental',
                        "Experimental"
                    )),
                    settings: ['application.experimental.*']
                },
                {
                    id: 'application/other',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'other',
                        "Other"
                    )),
                    settings: ['application.*']
                }
            ]
        },
        {
            id: 'security',
            label: ( localizeWithPath(
                'vs/workbench/contrib/preferences/browser/settingsLayout',
                'security',
                "Security"
            )),
            settings: ['security.*'],
            children: [
                {
                    id: 'security/workspace',
                    label: ( localizeWithPath(
                        'vs/workbench/contrib/preferences/browser/settingsLayout',
                        'workspace',
                        "Workspace"
                    )),
                    settings: ['security.workspace.*']
                }
            ]
        }
    ]
};
const knownAcronyms = ( new Set());
[
    'css',
    'html',
    'scss',
    'less',
    'json',
    'js',
    'ts',
    'ie',
    'id',
    'php',
    'scm',
].forEach(str => knownAcronyms.add(str));
const knownTermMappings = ( new Map());
knownTermMappings.set('power shell', 'PowerShell');
knownTermMappings.set('powershell', 'PowerShell');
knownTermMappings.set('javascript', 'JavaScript');
knownTermMappings.set('typescript', 'TypeScript');
export { getCommonlyUsedData, knownAcronyms, knownTermMappings, tocData };
