// vscode/vscode/src/vs/workbench/contrib/accessibility/browser/accessibilityConfiguration.js
import { __decorate, __param } from '../../../../../../../external/tslib/tslib.es6.js';
import { localizeWithPath } from '../../../../nls.js';
import { Extensions as Extensions$1 } from '../../../../platform/configuration/common/configurationRegistry.js';
import { Registry } from '../../../../platform/registry/common/platform.js';
import { RawContextKey } from '../../../../platform/contextkey/common/contextkey.js';
import { Extensions, workbenchConfigurationNodeBase } from '../../../common/configuration.js';
import { AccessibilitySignal } from '../../../../platform/accessibilitySignal/browser/accessibilitySignalService.js';
import { ISpeechService } from '../../speech/common/speechService.js';
import { Disposable } from '../../../../base/common/lifecycle.js';
import { Event } from '../../../../base/common/event.js';
import { soundFeatureBase } from '../../accessibilitySignals/browser/accessibilitySignal.contribution.js';
const accessibilityHelpIsShown = ( new RawContextKey('accessibilityHelpIsShown', false, true));
const accessibleViewIsShown = ( new RawContextKey('accessibleViewIsShown', false, true));
const accessibleViewSupportsNavigation = ( new RawContextKey('accessibleViewSupportsNavigation', false, true));
const accessibleViewVerbosityEnabled = ( new RawContextKey('accessibleViewVerbosityEnabled', false, true));
const accessibleViewGoToSymbolSupported = ( new RawContextKey('accessibleViewGoToSymbolSupported', false, true));
const accessibleViewOnLastLine = ( new RawContextKey('accessibleViewOnLastLine', false, true));
const accessibleViewCurrentProviderId = ( new RawContextKey('accessibleViewCurrentProviderId', undefined, undefined));
const baseVerbosityProperty = {
    type: 'boolean',
    default: true,
    tags: ['accessibility']
};
const markdownDeprecationMessage = ( localizeWithPath(
    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
    'accessibility.announcement.deprecationMessage',
    "This setting is deprecated. Use the `signals` settings instead."
));
const baseAlertProperty = {
    type: 'boolean',
    default: true,
    tags: ['accessibility'],
    markdownDeprecationMessage
};
const accessibilityConfigurationNodeBase = ( Object.freeze({
    id: 'accessibility',
    title: ( localizeWithPath(
        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
        'accessibilityConfigurationTitle',
        "Accessibility"
    )),
    type: 'object'
}));
const signalFeatureBase = {
    'type': 'object',
    'tags': ['accessibility'],
    additionalProperties: false,
    default: {
        sound: 'auto',
        announcement: 'auto'
    }
};
const announcementFeatureBase = {
    'type': 'string',
    'enum': ['auto', 'off'],
    'default': 'auto',
    'enumDescriptions': [
        ( localizeWithPath(
            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
            'announcement.enabled.auto',
            "Enable announcement, will only play when in screen reader optimized mode."
        )),
        ( localizeWithPath(
            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
            'announcement.enabled.off',
            "Disable announcement."
        ))
    ],
    tags: ['accessibility'],
};
const defaultNoAnnouncement = {
    'type': 'object',
    'tags': ['accessibility'],
    additionalProperties: false,
    'default': {
        'sound': 'auto',
    }
};
const configuration = {
    ...accessibilityConfigurationNodeBase,
    properties: {
        ["accessibility.verbosity.terminal" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.terminal.description',
                'Provide information about how to access the terminal accessibility help menu when the terminal is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.diffEditor" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.diffEditor.description',
                'Provide information about how to navigate changes in the diff editor when it is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.panelChat" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.chat.description',
                'Provide information about how to access the chat help menu when the chat input is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.inlineChat" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.interactiveEditor.description',
                'Provide information about how to access the inline editor chat accessibility help menu and alert with hints that describe how to use the feature when the input is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.inlineCompletions" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.inlineCompletions.description',
                'Provide information about how to access the inline completions hover and Accessible View.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.keybindingsEditor" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.keybindingsEditor.description',
                'Provide information about how to change a keybinding in the keybindings editor when a row is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.notebook" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.notebook',
                'Provide information about how to focus the cell container or inner editor when a notebook cell is focused.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.hover" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.hover',
                'Provide information about how to open the hover in an Accessible View.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.notification" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.notification',
                'Provide information about how to open the notification in an Accessible View.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.emptyEditorHint" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.emptyEditorHint',
                'Provide information about relevant actions in an empty text editor.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.verbosity.comments" ]: {
            description: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'verbosity.comments',
                'Provide information about actions that can be taken in the comment widget or in a file which contains comments.'
            )),
            ...baseVerbosityProperty
        },
        ["accessibility.alert.save" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.save',
                "Indicates when a file is saved. Also see {0}.",
                '`#audioCues.save#`'
            )),
            'enum': ['userGesture', 'always', 'never'],
            'default': 'always',
            'enumDescriptions': [
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.save.userGesture',
                    "Indicates when a file is saved via user gesture."
                )),
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.save.always',
                    "Indicates whenever is a file is saved, including auto save."
                )),
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.save.never',
                    "Never alerts."
                ))
            ],
            tags: ['accessibility'],
            markdownDeprecationMessage
        },
        ["accessibility.alert.clear" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.clear',
                "Indicates when a feature is cleared (for example, the terminal, Debug Console, or Output channel). Also see {0}.",
                '`#audioCues.clear#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.format" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.format',
                "Indicates when a file or notebook cell is formatted. Also see {0}.",
                '`#audioCues.format#`'
            )),
            'type': 'string',
            'enum': ['userGesture', 'always', 'never'],
            'default': 'always',
            'enumDescriptions': [
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.format.userGesture',
                    "Indicates when a file is formatted via user gesture."
                )),
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.format.always',
                    "Indicates whenever is a file is formatted, including auto save, on cell execution, and more."
                )),
                ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'announcement.format.never',
                    "Never alerts."
                ))
            ],
            tags: ['accessibility'],
            markdownDeprecationMessage
        },
        ["accessibility.alert.breakpoint" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.breakpoint',
                "Indicates when the debugger breaks. Also see {0}.",
                '`#audioCues.onDebugBreak#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.error" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.error',
                "Indicates when the active line has an error. Also see {0}.",
                '`#audioCues.lineHasError#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.warning" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.warning',
                "Indicates when the active line has a warning. Also see {0}.",
                '`#audioCues.lineHasWarning#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.foldedArea" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.foldedArea',
                "Indicates when the active line has a folded area that can be unfolded. Also see {0}.",
                '`#audioCues.lineHasFoldedArea#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.terminalQuickFix" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.terminalQuickFix',
                "Indicates when there is an available terminal quick fix. Also see {0}.",
                '`#audioCues.terminalQuickFix#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.terminalBell" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.terminalBell',
                "Indicates when the terminal bell is activated."
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.terminalCommandFailed" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.terminalCommandFailed',
                "Indicates when a terminal command fails (non-zero exit code). Also see {0}.",
                '`#audioCues.terminalCommandFailed#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.taskFailed" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.taskFailed',
                "Indicates when a task fails (non-zero exit code). Also see {0}.",
                '`#audioCues.taskFailed#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.taskCompleted" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.taskCompleted',
                "Indicates when a task completes successfully (zero exit code). Also see {0}.",
                '`#audioCues.taskCompleted#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.chatRequestSent" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.chatRequestSent',
                "Indicates when a chat request is sent. Also see {0}.",
                '`#audioCues.chatRequestSent#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.chatResponsePending" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.chatResponsePending',
                "Indicates when a chat response is pending. Also see {0}.",
                '`#audioCues.chatResponsePending#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.noInlayHints" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.noInlayHints',
                "Indicates when there are no inlay hints. Also see {0}.",
                '`#audioCues.noInlayHints#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.lineHasBreakpoint" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.lineHasBreakpoint',
                "Indicates when on a line with a breakpoint. Also see {0}.",
                '`#audioCues.lineHasBreakpoint#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.notebookCellCompleted" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.notebookCellCompleted',
                "Indicates when a notebook cell completes successfully. Also see {0}.",
                '`#audioCues.notebookCellCompleted#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.notebookCellFailed" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.notebookCellFailed',
                "Indicates when a notebook cell fails. Also see {0}.",
                '`#audioCues.notebookCellFailed#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.alert.onDebugBreak" ]: {
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'announcement.onDebugBreak',
                "Indicates when the debugger breaks. Also see {0}.",
                '`#audioCues.onDebugBreak#`'
            )),
            ...baseAlertProperty
        },
        ["accessibility.accessibleView.closeOnKeyPress" ]: {
            markdownDescription: ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'terminal.integrated.accessibleView.closeOnKeyPress',
                "On keypress, close the Accessible View and focus the element from which it was invoked."
            )),
            type: 'boolean',
            default: true
        },
        'accessibility.signals.sounds.volume': {
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.sounds.volume',
                "The volume of the sounds in percent (0-100)."
            )),
            'type': 'number',
            'minimum': 0,
            'maximum': 100,
            'default': 70,
            tags: ['accessibility']
        },
        'accessibility.signals.debouncePositionChanges': {
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.debouncePositionChanges',
                "Whether or not position changes should be debounced"
            )),
            'type': 'boolean',
            'default': false,
            tags: ['accessibility']
        },
        'accessibility.signals.lineHasBreakpoint': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.lineHasBreakpoint',
                "Plays a signal when the active line has a breakpoint."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasBreakpoint.sound',
                        "Plays a sound when the active line has a breakpoint."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasBreakpoint.announcement',
                        "Indicates when the active line has a breakpoint."
                    )),
                    ...announcementFeatureBase
                },
            },
        },
        'accessibility.signals.lineHasInlineSuggestion': {
            ...defaultNoAnnouncement,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.lineHasInlineSuggestion',
                "Indicates when the active line has an inline suggestion."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasInlineSuggestion.sound',
                        "Plays a sound when the active line has an inline suggestion."
                    )),
                    ...soundFeatureBase,
                    'default': 'off'
                }
            }
        },
        'accessibility.signals.lineHasError': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.lineHasError',
                "Indicates when the active line has an error."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasError.sound',
                        "Plays a sound when the active line has an error."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasError.announcement',
                        "Indicates when the active line has an error."
                    )),
                    ...announcementFeatureBase,
                    default: 'off'
                },
            },
        },
        'accessibility.signals.lineHasFoldedArea': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.lineHasFoldedArea',
                "Indicates when the active line has a folded area that can be unfolded."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasFoldedArea.sound',
                        "Plays a sound when the active line has a folded area that can be unfolded."
                    )),
                    ...soundFeatureBase,
                    default: 'off'
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasFoldedArea.announcement',
                        "Indicates when the active line has a folded area that can be unfolded."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.lineHasWarning': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.lineHasWarning',
                "Plays a signal when the active line has a warning."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasWarning.sound',
                        "Plays a sound when the active line has a warning."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.lineHasWarning.announcement',
                        "Indicates when the active line has a warning."
                    )),
                    ...announcementFeatureBase,
                    default: 'off'
                },
            },
        },
        'accessibility.signals.onDebugBreak': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.onDebugBreak',
                "Plays a signal when the debugger stopped on a breakpoint."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.onDebugBreak.sound',
                        "Plays a sound when the debugger stopped on a breakpoint."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.onDebugBreak.announcement',
                        "Indicates when the debugger stopped on a breakpoint."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.noInlayHints': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.noInlayHints',
                "Plays a signal when trying to read a line with inlay hints that has no inlay hints."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.noInlayHints.sound',
                        "Plays a sound when trying to read a line with inlay hints that has no inlay hints."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.noInlayHints.announcement',
                        "Indicates when trying to read a line with inlay hints that has no inlay hints."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.taskCompleted': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.taskCompleted',
                "Plays a signal when a task is completed."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.taskCompleted.sound',
                        "Plays a sound when a task is completed."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.taskCompleted.announcement',
                        "Indicates when a task is completed."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.taskFailed': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.taskFailed',
                "Plays a signal when a task fails (non-zero exit code)."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.taskFailed.sound',
                        "Plays a sound when a task fails (non-zero exit code)."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.taskFailed.announcement',
                        "Indicates when a task fails (non-zero exit code)."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.terminalCommandFailed': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.terminalCommandFailed',
                "Plays a signal when a terminal command fails (non-zero exit code)."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalCommandFailed.sound',
                        "Plays a sound when a terminal command fails (non-zero exit code)."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalCommandFailed.announcement',
                        "Indicates when a terminal command fails (non-zero exit code)."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.terminalQuickFix': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.terminalQuickFix',
                "Plays a signal when terminal Quick Fixes are available."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalQuickFix.sound',
                        "Plays a sound when terminal Quick Fixes are available."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalQuickFix.announcement',
                        "Indicates when terminal Quick Fixes are available."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.terminalBell': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.terminalBell',
                "Plays a signal when the terminal bell is ringing."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalBell.sound',
                        "Plays a sound when the terminal bell is ringing."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.terminalBell.announcement',
                        "Indicates when the terminal bell is ringing."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.diffLineInserted': {
            ...defaultNoAnnouncement,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.diffLineInserted',
                "Indicates when the focus moves to an inserted line in Accessible Diff Viewer mode or to the next/previous change."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.sound',
                        "Plays a sound when the focus moves to an inserted line in Accessible Diff Viewer mode or to the next/previous change."
                    )),
                    ...soundFeatureBase
                }
            }
        },
        'accessibility.signals.diffLineModified': {
            ...defaultNoAnnouncement,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.diffLineModified',
                "Indicates when the focus moves to an modified line in Accessible Diff Viewer mode or to the next/previous change."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.diffLineModified.sound',
                        "Plays a sound when the focus moves to a modified line in Accessible Diff Viewer mode or to the next/previous change."
                    )),
                    ...soundFeatureBase
                }
            }
        },
        'accessibility.signals.diffLineDeleted': {
            ...defaultNoAnnouncement,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.diffLineDeleted',
                "Indicates when the focus moves to an deleted line in Accessible Diff Viewer mode or to the next/previous change."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.diffLineDeleted.sound',
                        "Plays a sound when the focus moves to an deleted line in Accessible Diff Viewer mode or to the next/previous change."
                    )),
                    ...soundFeatureBase
                }
            }
        },
        'accessibility.signals.notebookCellCompleted': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.notebookCellCompleted',
                "Plays a signal when a notebook cell execution is successfully completed."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.notebookCellCompleted.sound',
                        "Plays a sound when a notebook cell execution is successfully completed."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.notebookCellCompleted.announcement',
                        "Indicates when a notebook cell execution is successfully completed."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.notebookCellFailed': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.notebookCellFailed',
                "Plays a signal when a notebook cell execution fails."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.notebookCellFailed.sound',
                        "Plays a sound when a notebook cell execution fails."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.notebookCellFailed.announcement',
                        "Indicates when a notebook cell execution fails."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.chatRequestSent': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.chatRequestSent',
                "Plays a signal when a chat request is made."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.chatRequestSent.sound',
                        "Plays a sound when a chat request is made."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.chatRequestSent.announcement',
                        "Indicates when a chat request is made."
                    )),
                    ...announcementFeatureBase
                },
            }
        },
        'accessibility.signals.chatResponsePending': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.chatResponsePending',
                "Plays a signal on loop while the response is pending."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.chatResponsePending.sound',
                        "Plays a sound on loop while the response is pending."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.chatResponsePending.announcement',
                        "Alerts on loop while the response is pending."
                    )),
                    ...announcementFeatureBase
                },
            },
        },
        'accessibility.signals.chatResponseReceived': {
            ...defaultNoAnnouncement,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.chatResponseReceived',
                "Indicates when the response has been received."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.chatResponseReceived.sound',
                        "Plays a sound on loop while the response has been received."
                    )),
                    ...soundFeatureBase
                },
            }
        },
        'accessibility.signals.clear': {
            ...signalFeatureBase,
            'description': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.clear',
                "Plays a signal when a feature is cleared (for example, the terminal, Debug Console, or Output channel)."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.clear.sound',
                        "Plays a sound when a feature is cleared."
                    )),
                    ...soundFeatureBase
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.clear.announcement',
                        "Indicates when a feature is cleared."
                    )),
                    ...announcementFeatureBase
                },
            },
        },
        'accessibility.signals.save': {
            'type': 'object',
            'tags': ['accessibility'],
            additionalProperties: false,
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.save',
                "Plays a signal when a file is saved."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.save.sound',
                        "Plays a sound when a file is saved."
                    )),
                    'type': 'string',
                    'enum': ['userGesture', 'always', 'never'],
                    'default': 'never',
                    'enumDescriptions': [
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.sound.userGesture',
                            "Plays the audio cue when a user explicitly saves a file."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.sound.always',
                            "Plays the audio cue whenever a file is saved, including auto save."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.sound.never',
                            "Never plays the audio cue."
                        ))
                    ],
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.save.announcement',
                        "Indicates when a file is saved."
                    )),
                    'type': 'string',
                    'enum': ['userGesture', 'always', 'never'],
                    'default': 'never',
                    'enumDescriptions': [
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.announcement.userGesture',
                            "Announces when a user explicitly saves a file."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.announcement.always',
                            "Announces whenever a file is saved, including auto save."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.save.announcement.never',
                            "Never plays the audio cue."
                        ))
                    ],
                },
            },
            default: {
                'sound': 'never',
                'announcement': 'never'
            }
        },
        'accessibility.signals.format': {
            'type': 'object',
            'tags': ['accessibility'],
            additionalProperties: false,
            'markdownDescription': ( localizeWithPath(
                'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                'accessibility.signals.format',
                "Plays a signal when a file or notebook is formatted."
            )),
            'properties': {
                'sound': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.format.sound',
                        "Plays a sound when a file or notebook is formatted."
                    )),
                    'type': 'string',
                    'enum': ['userGesture', 'always', 'never'],
                    'default': 'never',
                    'enumDescriptions': [
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.userGesture',
                            "Plays the audio cue when a user explicitly formats a file."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.always',
                            "Plays the audio cue whenever a file is formatted, including if it is set to format on save, type, or, paste, or run of a cell."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.never',
                            "Never plays the audio cue."
                        ))
                    ],
                },
                'announcement': {
                    'description': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'accessibility.signals.format.announcement',
                        "Indicates when a file or notebook is formatted."
                    )),
                    'type': 'string',
                    'enum': ['userGesture', 'always', 'never'],
                    'default': 'never',
                    'enumDescriptions': [
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.announcement.userGesture',
                            "Announceswhen a user explicitly formats a file."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.announcement.always',
                            "Announces whenever a file is formatted, including if it is set to format on save, type, or, paste, or run of a cell."
                        )),
                        ( localizeWithPath(
                            'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                            'accessibility.signals.format.announcement.never',
                            "Never announces."
                        ))
                    ],
                },
            },
            default: {
                'sound': 'never',
                'announcement': 'never'
            }
        },
    }
};
function registerAccessibilityConfiguration() {
    const registry = ( Registry.as(Extensions$1.Configuration));
    registry.registerConfiguration(configuration);
    registry.registerConfiguration({
        ...workbenchConfigurationNodeBase,
        properties: {
            ["accessibility.dimUnfocused.enabled" ]: {
                description: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'dimUnfocusedEnabled',
                    'Whether to dim unfocused editors and terminals, which makes it more clear where typed input will go to. This works with the majority of editors with the notable exceptions of those that utilize iframes like notebooks and extension webview editors.'
                )),
                type: 'boolean',
                default: false,
                tags: ['accessibility'],
                scope: 1 ,
            },
            ["accessibility.dimUnfocused.opacity" ]: {
                markdownDescription: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'dimUnfocusedOpacity',
                    'The opacity fraction (0.2 to 1.0) to use for unfocused editors and terminals. This will only take effect when {0} is enabled.',
                    `\`#${"accessibility.dimUnfocused.enabled" }#\``
                )),
                type: 'number',
                minimum: 0.2 ,
                maximum: 1 ,
                default: 0.75 ,
                tags: ['accessibility'],
                scope: 1 ,
            },
            ["accessibility.hideAccessibleView" ]: {
                description: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'accessibility.hideAccessibleView',
                    "Controls whether the Accessible View is hidden."
                )),
                type: 'boolean',
                default: false,
                tags: ['accessibility']
            }
        }
    });
}
const SpeechTimeoutDefault = 1200;
const SpeechLanguageDefault = 'en-US';
let DynamicSpeechAccessibilityConfiguration = class DynamicSpeechAccessibilityConfiguration extends Disposable {
    static { this.ID = 'workbench.contrib.dynamicSpeechAccessibilityConfiguration'; }
    constructor(speechService) {
        super();
        this.speechService = speechService;
        this._register(Event.runAndSubscribe(speechService.onDidRegisterSpeechProvider, () => this.updateConfiguration()));
    }
    updateConfiguration() {
        if (!this.speechService.hasSpeechProvider) {
            return;
        }
        const languages = this.getLanguages();
        const languagesSorted = ( Object.keys(languages)).sort((langA, langB) => {
            return languages[langA].name.localeCompare(languages[langB].name);
        });
        const registry = ( Registry.as(Extensions$1.Configuration));
        registry.registerConfiguration({
            ...accessibilityConfigurationNodeBase,
            properties: {
                ["accessibility.voice.speechTimeout" ]: {
                    'markdownDescription': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'voice.speechTimeout',
                        "The duration in milliseconds that voice speech recognition remains active after you stop speaking. For example in a chat session, the transcribed text is submitted automatically after the timeout is met. Set to `0` to disable this feature."
                    )),
                    'type': 'number',
                    'default': SpeechTimeoutDefault,
                    'minimum': 0,
                    'tags': ['accessibility']
                },
                ["accessibility.voice.speechLanguage" ]: {
                    'markdownDescription': ( localizeWithPath(
                        'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                        'voice.speechLanguage',
                        "The language that voice speech recognition should recognize."
                    )),
                    'type': 'string',
                    'enum': languagesSorted,
                    'default': SpeechLanguageDefault,
                    'tags': ['accessibility'],
                    'enumDescriptions': ( languagesSorted.map(key => languages[key].name)),
                    'enumItemLabels': ( languagesSorted.map(key => languages[key].name))
                }
            }
        });
    }
    getLanguages() {
        return {
            ['da-DK']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.da-DK',
                    "Danish (Denmark)"
                ))
            },
            ['de-DE']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.de-DE',
                    "German (Germany)"
                ))
            },
            ['en-AU']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-AU',
                    "English (Australia)"
                ))
            },
            ['en-CA']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-CA',
                    "English (Canada)"
                ))
            },
            ['en-GB']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-GB',
                    "English (United Kingdom)"
                ))
            },
            ['en-IE']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-IE',
                    "English (Ireland)"
                ))
            },
            ['en-IN']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-IN',
                    "English (India)"
                ))
            },
            ['en-NZ']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-NZ',
                    "English (New Zealand)"
                ))
            },
            [SpeechLanguageDefault]: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.en-US',
                    "English (United States)"
                ))
            },
            ['es-ES']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.es-ES',
                    "Spanish (Spain)"
                ))
            },
            ['es-MX']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.es-MX',
                    "Spanish (Mexico)"
                ))
            },
            ['fr-CA']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.fr-CA',
                    "French (Canada)"
                ))
            },
            ['fr-FR']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.fr-FR',
                    "French (France)"
                ))
            },
            ['hi-IN']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.hi-IN',
                    "Hindi (India)"
                ))
            },
            ['it-IT']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.it-IT',
                    "Italian (Italy)"
                ))
            },
            ['ja-JP']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.ja-JP',
                    "Japanese (Japan)"
                ))
            },
            ['ko-KR']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.ko-KR',
                    "Korean (South Korea)"
                ))
            },
            ['nl-NL']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.nl-NL',
                    "Dutch (Netherlands)"
                ))
            },
            ['pt-PT']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.pt-PT',
                    "Portuguese (Portugal)"
                ))
            },
            ['pt-BR']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.pt-BR',
                    "Portuguese (Brazil)"
                ))
            },
            ['ru-RU']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.ru-RU',
                    "Russian (Russia)"
                ))
            },
            ['sv-SE']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.sv-SE',
                    "Swedish (Sweden)"
                ))
            },
            ['tr-TR']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.tr-TR',
                    "Turkish (Turkey)"
                ))
            },
            ['zh-CN']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.zh-CN',
                    "Chinese (Simplified, China)"
                ))
            },
            ['zh-HK']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.zh-HK',
                    "Chinese (Traditional, Hong Kong)"
                ))
            },
            ['zh-TW']: {
                name: ( localizeWithPath(
                    'vs/workbench/contrib/accessibility/browser/accessibilityConfiguration',
                    'speechLanguage.zh-TW',
                    "Chinese (Traditional, Taiwan)"
                ))
            }
        };
    }
};
DynamicSpeechAccessibilityConfiguration = ( __decorate([
    ( __param(0, ISpeechService))
], DynamicSpeechAccessibilityConfiguration));
( Registry.as(Extensions.ConfigurationMigration))
    .registerConfigurationMigrations([{
        key: 'audioCues.volume',
        migrateFn: (value, accessor) => {
            return [
                ['accessibility.signals.sounds.volume', { value }],
                ['audioCues.volume', { value: undefined }]
            ];
        }
    }]);
( Registry.as(Extensions.ConfigurationMigration))
    .registerConfigurationMigrations([{
        key: 'audioCues.debouncePositionChanges',
        migrateFn: (value, accessor) => {
            return [
                ['accessibility.signals.debouncePositionChanges', { value }],
                ['audioCues.debouncePositionChanges', { value: undefined }]
            ];
        }
    }]);
( Registry.as(Extensions.ConfigurationMigration))
    .registerConfigurationMigrations(( AccessibilitySignal.allAccessibilitySignals.map(item => ({
    key: item.legacySoundSettingsKey,
    migrateFn: (sound, accessor) => {
        const configurationKeyValuePairs = [];
        const legacyAnnouncementSettingsKey = item.legacyAnnouncementSettingsKey;
        let announcement;
        if (legacyAnnouncementSettingsKey) {
            announcement = accessor(legacyAnnouncementSettingsKey) ?? undefined;
            if (announcement !== undefined && typeof announcement !== 'string') {
                announcement = announcement ? 'auto' : 'off';
            }
        }
        configurationKeyValuePairs.push([`${item.legacySoundSettingsKey}`, { value: undefined }]);
        configurationKeyValuePairs.push([`${item.settingsKey}`, { value: announcement !== undefined ? { announcement, sound } : { sound } }]);
        return configurationKeyValuePairs;
    }
}))));
( Registry.as(Extensions.ConfigurationMigration))
    .registerConfigurationMigrations(( AccessibilitySignal.allAccessibilitySignals.filter(i => !!i.legacyAnnouncementSettingsKey).map(item => ({
    key: item.legacyAnnouncementSettingsKey,
    migrateFn: (announcement, accessor) => {
        const configurationKeyValuePairs = [];
        const sound = accessor(item.settingsKey)?.sound || accessor(item.legacySoundSettingsKey);
        if (announcement !== undefined && typeof announcement !== 'string') {
            announcement = announcement ? 'auto' : 'off';
        }
        configurationKeyValuePairs.push([`${item.settingsKey}`, { value: announcement !== undefined ? { announcement, sound } : { sound } }]);
        configurationKeyValuePairs.push([`${item.legacyAnnouncementSettingsKey}`, { value: undefined }]);
        configurationKeyValuePairs.push([`${item.legacySoundSettingsKey}`, { value: undefined }]);
        return configurationKeyValuePairs;
    }
}))));
export { DynamicSpeechAccessibilityConfiguration, SpeechTimeoutDefault, accessibilityConfigurationNodeBase, accessibilityHelpIsShown, accessibleViewCurrentProviderId, accessibleViewGoToSymbolSupported, accessibleViewIsShown, accessibleViewOnLastLine, accessibleViewSupportsNavigation, accessibleViewVerbosityEnabled, announcementFeatureBase, registerAccessibilityConfiguration };
