// vscode/vscode/src/vs/workbench/contrib/comments/common/commentModel.js
class CommentNode {
    constructor(owner, threadId, resource, comment, range, threadState) {
        this.replies = [];
        this.owner = owner;
        this.threadId = threadId;
        this.comment = comment;
        this.resource = resource;
        this.range = range;
        this.isRoot = false;
        this.threadState = threadState;
    }
    hasReply() {
        return this.replies && this.replies.length !== 0;
    }
}
class ResourceWithCommentThreads {
    constructor(owner, resource, commentThreads) {
        this.owner = owner;
        this.id = ( resource.toString());
        this.resource = resource;
        this.commentThreads = ( commentThreads.filter(thread => thread.comments && thread.comments.length).map(
            thread => ResourceWithCommentThreads.createCommentNode(owner, resource, thread)
        ));
    }
    static createCommentNode(owner, resource, commentThread) {
        const { threadId, comments, range } = commentThread;
        const commentNodes = ( comments.map(comment => ( new CommentNode(owner, threadId, resource, comment, range, commentThread.state))));
        if (commentNodes.length > 1) {
            commentNodes[0].replies = commentNodes.slice(1, commentNodes.length);
        }
        commentNodes[0].isRoot = true;
        return commentNodes[0];
    }
}
export { CommentNode, ResourceWithCommentThreads };
