// vscode/vscode/src/vs/workbench/services/configurationResolver/common/configurationResolverSchema.js
import { localizeWithPath } from '../../../../nls.js';
const idDescription = ( localizeWithPath(
    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
    'JsonSchema.input.id',
    "The input's id is used to associate an input with a variable of the form ${input:id}."
));
const typeDescription = ( localizeWithPath(
    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
    'JsonSchema.input.type',
    "The type of user input prompt to use."
));
const descriptionDescription = ( localizeWithPath(
    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
    'JsonSchema.input.description',
    "The description is shown when the user is prompted for input."
));
const defaultDescription = ( localizeWithPath(
    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
    'JsonSchema.input.default',
    "The default value for the input."
));
const inputsSchema = {
    definitions: {
        inputs: {
            type: 'array',
            description: ( localizeWithPath(
                'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                'JsonSchema.inputs',
                'User inputs. Used for defining user input prompts, such as free string input or a choice from several options.'
            )),
            items: {
                oneOf: [
                    {
                        type: 'object',
                        required: ['id', 'type', 'description'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['promptString'],
                                enumDescriptions: [
                                    ( localizeWithPath(
                                        'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                        'JsonSchema.input.type.promptString',
                                        "The 'promptString' type opens an input box to ask the user for input."
                                    )),
                                ]
                            },
                            description: {
                                type: 'string',
                                description: descriptionDescription
                            },
                            default: {
                                type: 'string',
                                description: defaultDescription
                            },
                            password: {
                                type: 'boolean',
                                description: ( localizeWithPath(
                                    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                    'JsonSchema.input.password',
                                    "Controls if a password input is shown. Password input hides the typed text."
                                )),
                            },
                        }
                    },
                    {
                        type: 'object',
                        required: ['id', 'type', 'description', 'options'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['pickString'],
                                enumDescriptions: [
                                    ( localizeWithPath(
                                        'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                        'JsonSchema.input.type.pickString',
                                        "The 'pickString' type shows a selection list."
                                    )),
                                ]
                            },
                            description: {
                                type: 'string',
                                description: descriptionDescription
                            },
                            default: {
                                type: 'string',
                                description: defaultDescription
                            },
                            options: {
                                type: 'array',
                                description: ( localizeWithPath(
                                    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                    'JsonSchema.input.options',
                                    "An array of strings that defines the options for a quick pick."
                                )),
                                items: {
                                    oneOf: [
                                        {
                                            type: 'string'
                                        },
                                        {
                                            type: 'object',
                                            required: ['value'],
                                            additionalProperties: false,
                                            properties: {
                                                label: {
                                                    type: 'string',
                                                    description: ( localizeWithPath(
                                                        'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                                        'JsonSchema.input.pickString.optionLabel',
                                                        "Label for the option."
                                                    ))
                                                },
                                                value: {
                                                    type: 'string',
                                                    description: ( localizeWithPath(
                                                        'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                                        'JsonSchema.input.pickString.optionValue',
                                                        "Value for the option."
                                                    ))
                                                }
                                            }
                                        }
                                    ]
                                }
                            }
                        }
                    },
                    {
                        type: 'object',
                        required: ['id', 'type', 'command'],
                        additionalProperties: false,
                        properties: {
                            id: {
                                type: 'string',
                                description: idDescription
                            },
                            type: {
                                type: 'string',
                                description: typeDescription,
                                enum: ['command'],
                                enumDescriptions: [
                                    ( localizeWithPath(
                                        'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                        'JsonSchema.input.type.command',
                                        "The 'command' type executes a command."
                                    )),
                                ]
                            },
                            command: {
                                type: 'string',
                                description: ( localizeWithPath(
                                    'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                    'JsonSchema.input.command.command',
                                    "The command to execute for this input variable."
                                ))
                            },
                            args: {
                                oneOf: [
                                    {
                                        type: 'object',
                                        description: ( localizeWithPath(
                                            'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                            'JsonSchema.input.command.args',
                                            "Optional arguments passed to the command."
                                        ))
                                    },
                                    {
                                        type: 'array',
                                        description: ( localizeWithPath(
                                            'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                            'JsonSchema.input.command.args',
                                            "Optional arguments passed to the command."
                                        ))
                                    },
                                    {
                                        type: 'string',
                                        description: ( localizeWithPath(
                                            'vs/workbench/services/configurationResolver/common/configurationResolverSchema',
                                            'JsonSchema.input.command.args',
                                            "Optional arguments passed to the command."
                                        ))
                                    }
                                ]
                            }
                        }
                    }
                ]
            }
        }
    }
};
export { inputsSchema };
